{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SNP filtering using vcftools   \n",
    "\n",
    "\n",
    "use preferred conda env  \n",
    "**Packages needed**: vcftools, bgzip, tabix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Things we filter on:  \n",
    "\n",
    "#### Individuals:  \n",
    "- **Coverage**: Also can be thought of as **depth**.See 3mapping. Calculated on bam files. Average read count per locus per individual. \n",
    "- **Missing**: Proportion of missing data allowed across all loci for individual. Common and high in GBS/RADseq data. Kinda an issue all around. Many methods, including PCA (all ordination methods), require a complete matrix with no missing data. Additionally, PCA will cluster by missing data with individuals with higher missing data clustering closer to the center and get this \"fan\" effect. Can be the same for coverage too. This (among other reasons) is why people use a variance-covariance matrix of genetic data to do ordinations. Other methods involve imputation. This can be fancy and use phased haplotype data OR simply, when you z-score, (g - mean(g))/sd(g), your genotype data across each locus, you make all missing data equal to 0 or Mean (i.e., the global allele frequency). There's more to this standardization, see *Patterson et al. 2006* (https://dx.plos.org/10.1371/journal.pgen.0020190) for more info. See PCAsim_ex in examples directory for showing all these issues.\n",
    "    - (additional) This is another reason to use entropy. Entropy is a hierarchical bayesian model so it gets an updated genotype estimate for each missing value based on genotype likelihoods across loci, individuals, and the allele frequency of the cluster/deme that individual assigns to.   \n",
    "    \n",
    "#### Loci:  \n",
    "- **Biallelic**: Only keep biallelic SNPs. Multiallelic SNPs are rare at the time scale we work (Citation??) and also,  mathematical nightmare and we have enough data so just ignore. Everyone does unless deep time phylogenetics. \n",
    "- **thin**: Keeps one locus within a specified range. Not 100% how it decides with one to keep. I think it's on quality or depth. This is a necessary step as loci in close physical are prone to sequencing error and linkage disequalibrium (LD) confounds many different population genetic parameters. For *de novo* reference assemblies, we thin to 100 as contigs/reads are ~92 bp in length. This keeps one locus per contig to control for LD and sequencing error, which is really common in pop gen and necessary for many analyses.   \n",
    "- **max-missing** = max proportion of missing data per locus  \n",
    "- **MAF** = minor allele frequency. Proportion of individuals a alternate allele needs to be present in order for that locus to be kept as a SNP. (e.g. maf = 0.02 for 250 individuals means that an alternate allele needs to be present in at least 5 individuals to be kept) Many papers have shown this is a big issue in clustering and demography (Citation). We do this a second time near the end if we removed individuals during missing data filtering.   \n",
    "- **Mean Depth**: Average allelic depth or read depth per locus. Too low could be sequencing error, too high could be PCR replication artifact (Citation).    \n",
    "- **Qual**: Locus Quality. You can look up the math. Usually above 20-30 is good but given our coverage and number of individuals, we can usually go way higher.     \n",
    "- **Fis**: Inbreeding coefficient. This is a contentous topic. This has to do with paralogs or paralogous loci. This is where loci map to multiple regions of the genome. Issues in highly repeative genomes. Usually leads to an excess of heterozygotes. Filtering on negative Fis can help. See these two McKinney papers (https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12763, https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12613). Katie and others in the lab use his package called HDPlot to deal with this.   \n",
    "\n",
    "\n",
    "**See methods_ex.Rmd in examples directory for text** \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {}
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import ipyparallel as ipp\n",
    "import os\n",
    "from os import environ\n",
    "import gzip\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import glob\n",
    "import re\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools = \"vcftools\"\n",
    "bgzip = \"bgzip\"\n",
    "tabix = \"tabix\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "root = '/data/gpfs/assoc/denovo/PHHA/SNPcall/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/PHHA/SNPcall\n"
     ]
    }
   ],
   "source": [
    "cd $root"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### cp vcf.gz from vcf_dir to filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp vcf/PHHA.vcf.gz filtering/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = os.path.join(root,\"filtering\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/PHHA/SNPcall/filtering\n"
     ]
    }
   ],
   "source": [
    "cd $analysis_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### count snps in zipped vcf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/PHHA.vcf.gz'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file = os.path.join(analysis_dir, \"PHHA.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\r\n"
     ]
    }
   ],
   "source": [
    "!zcat $vcf_file | grep -v '#' | wc -l "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keep only biallelic loci as first step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/PHHA.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--out PHHA.biallelic\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 248 out of 248 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 1334848 out of a possible 1582788 Sites\n",
      "Run Time = 657.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $'PHHA.biallelic'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_biallelic = os.path.join(analysis_dir, \"PHHA.biallelic.recode.vcf\")\n",
    "vcf_biallelic_gz = vcf_biallelic + '.gz'\n",
    "!$bgzip -c $vcf_biallelic > {vcf_biallelic_gz}\n",
    "!$tabix {vcf_biallelic_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove by MAF, missing data, and thin   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/PHHA.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--thin 100\n",
      "\t--max-missing 0.6\n",
      "\t--out PHHA_miss60_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 248 out of 248 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 42386 out of a possible 1334848 Sites\n",
      "Run Time = 79.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.6 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'PHHA_miss60_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/PHHA.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--thin 100\n",
      "\t--max-missing 0.7\n",
      "\t--out PHHA_miss70_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 248 out of 248 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 34737 out of a possible 1334848 Sites\n",
      "Run Time = 75.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.7 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'PHHA_miss70_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"PHHA_miss70_thin100_MAF2.recode.vcf\"\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove individuals with too much missing data (bad_indv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf PHHA_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--missing-indv\n",
      "\t--out PHHA_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 248 out of 248 Individuals\n",
      "Outputting Individual Missingness\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz --out $vcf_filtered_gz --missing-indv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_imiss(vcf_file):\n",
    "    imiss_file = !ls {vcf_file}.imiss\n",
    "    imiss_df = pd.read_csv(imiss_file[0], sep=\"\\t\")\n",
    "    imiss_df.index = imiss_df.INDV\n",
    "    return imiss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPES_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INDV</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PH_AS_10</th>\n",
       "      <td>PH_AS_10</td>\n",
       "      <td>34737</td>\n",
       "      <td>0</td>\n",
       "      <td>1467</td>\n",
       "      <td>0.042232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_11</th>\n",
       "      <td>PH_AS_11</td>\n",
       "      <td>34737</td>\n",
       "      <td>0</td>\n",
       "      <td>3811</td>\n",
       "      <td>0.109710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_12</th>\n",
       "      <td>PH_AS_12</td>\n",
       "      <td>34737</td>\n",
       "      <td>0</td>\n",
       "      <td>2870</td>\n",
       "      <td>0.082621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_1</th>\n",
       "      <td>PH_AS_1</td>\n",
       "      <td>34737</td>\n",
       "      <td>0</td>\n",
       "      <td>3023</td>\n",
       "      <td>0.087025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_2</th>\n",
       "      <td>PH_AS_2</td>\n",
       "      <td>34737</td>\n",
       "      <td>0</td>\n",
       "      <td>874</td>\n",
       "      <td>0.025160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              INDV  N_DATA  N_GENOTYPES_FILTERED  N_MISS    F_MISS\n",
       "INDV                                                              \n",
       "PH_AS_10  PH_AS_10   34737                     0    1467  0.042232\n",
       "PH_AS_11  PH_AS_11   34737                     0    3811  0.109710\n",
       "PH_AS_12  PH_AS_12   34737                     0    2870  0.082621\n",
       "PH_AS_1    PH_AS_1   34737                     0    3023  0.087025\n",
       "PH_AS_2    PH_AS_2   34737                     0     874  0.025160"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df = get_imiss(vcf_filtered_gz)\n",
    "imiss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    248.000000\n",
       "mean       0.125917\n",
       "std        0.088792\n",
       "min        0.015833\n",
       "25%        0.079800\n",
       "50%        0.098944\n",
       "75%        0.158412\n",
       "max        0.989953\n",
       "Name: F_MISS, dtype: float64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df.F_MISS.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Below is just allows you to input a cutoff and see how many individuals you lose, keeping individuals with at least 60% data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(248, 3, 1, 14)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(imiss_df),len(imiss_df[imiss_df.F_MISS >= .4]),len(imiss_df[imiss_df.F_MISS >= .5]), len(imiss_df[imiss_df.F_MISS >= .25])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_indv = imiss_df.INDV[imiss_df.F_MISS >= .4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"bad_indv.txt\"), \"w\") as o:\n",
    "    o.write(\"INDV\\n\")\n",
    "    for elem in bad_indv.index:\n",
    "        o.write(\"%s\\n\" % elem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf PHHA_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--remove /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/bad_indv.txt\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Excluding individuals in 'exclude' list\n",
      "After filtering, kept 245 out of 248 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 18.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--remove {os.path.join(analysis_dir, \"bad_indv.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"snps_indv_removed\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_bad_remove = os.path.join(analysis_dir, \"snps_indv_removed.recode.vcf\")\n",
    "vcf_bad_remove_gz = vcf_bad_remove + \".gz\"\n",
    "!$bgzip -c {vcf_bad_remove} > {vcf_bad_remove_gz}\n",
    "!$tabix {vcf_bad_remove_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter snps further \n",
    "this needs to be done after removing individuals   \n",
    "\n",
    "### This uses vcftools to get some stats to summarize and make decisions with later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 3.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 3.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Quality for Each Site\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 34737 out of a possible 34737 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_bad_remove_gz)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions that will calculate various metrics we use to filter   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    \n",
    "    hardy_files = !ls {filedir}/{prefix}.hwe\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_files = !ls {filedir}/{prefix}.l* | grep -v log\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    frq_files = !ls {filedir}/{prefix}.frq* | grep -v count\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_bad_remove_gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:45: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/pandas/core/indexing.py:822: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=6, step=1)\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(analysis_dir,'snps_indv_removed.recode.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>...</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_23-44</th>\n",
       "      <td>13650</td>\n",
       "      <td>1167216</td>\n",
       "      <td>55.71430</td>\n",
       "      <td>1666.87000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.012245</td>\n",
       "      <td>507.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.497843</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.635577</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.956612</td>\n",
       "      <td>0.043388</td>\n",
       "      <td>0.043388</td>\n",
       "      <td>-0.043197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_27-26</th>\n",
       "      <td>603</td>\n",
       "      <td>2977</td>\n",
       "      <td>2.46122</td>\n",
       "      <td>6.11837</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>114</td>\n",
       "      <td>0.232653</td>\n",
       "      <td>359.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1.517900</td>\n",
       "      <td>6.176162e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.259856</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.917553</td>\n",
       "      <td>0.082447</td>\n",
       "      <td>0.082447</td>\n",
       "      <td>-0.086957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_103-30</th>\n",
       "      <td>59707</td>\n",
       "      <td>14552605</td>\n",
       "      <td>243.70200</td>\n",
       "      <td>7.73463</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>977.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>245.000000</td>\n",
       "      <td>3.521762e-11</td>\n",
       "      <td>3.521762e-11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.979592</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_148-62</th>\n",
       "      <td>60976</td>\n",
       "      <td>15176182</td>\n",
       "      <td>248.88200</td>\n",
       "      <td>1.53921</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>213.330700</td>\n",
       "      <td>2.172593e-22</td>\n",
       "      <td>2.172593e-22</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.934694</td>\n",
       "      <td>0.065306</td>\n",
       "      <td>0.065306</td>\n",
       "      <td>0.933270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_152-75</th>\n",
       "      <td>58538</td>\n",
       "      <td>13991770</td>\n",
       "      <td>238.93100</td>\n",
       "      <td>21.51570</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>999.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>229.337900</td>\n",
       "      <td>9.782448e-25</td>\n",
       "      <td>9.782448e-25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.932653</td>\n",
       "      <td>0.067347</td>\n",
       "      <td>0.067347</td>\n",
       "      <td>0.967575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH   VAR_DEPTH  N_DATA  \\\n",
       "dDocent_Contig_23-44       13650      1167216    55.71430  1666.87000     490   \n",
       "dDocent_Contig_27-26         603         2977     2.46122     6.11837     490   \n",
       "dDocent_Contig_103-30      59707     14552605   243.70200     7.73463     490   \n",
       "dDocent_Contig_148-62      60976     15176182   248.88200     1.53921     490   \n",
       "dDocent_Contig_152-75      58538     13991770   238.93100    21.51570     490   \n",
       "\n",
       "                       N_GENOTYPE_FILTERED  N_MISS    F_MISS   QUAL N_ALLELES  \\\n",
       "dDocent_Contig_23-44                     0       6  0.012245  507.0         2   \n",
       "dDocent_Contig_27-26                     0     114  0.232653  359.0         2   \n",
       "dDocent_Contig_103-30                    0       0  0.000000  977.0         2   \n",
       "dDocent_Contig_148-62                    0       0  0.000000  999.0         2   \n",
       "dDocent_Contig_152-75                    0       0  0.000000  999.0         2   \n",
       "\n",
       "                       ...   ChiSq_HWE         P_HWE P_HET_DEFICIT  \\\n",
       "dDocent_Contig_23-44   ...    0.497843  1.000000e+00  1.000000e+00   \n",
       "dDocent_Contig_27-26   ...    1.517900  6.176162e-01  1.000000e+00   \n",
       "dDocent_Contig_103-30  ...  245.000000  3.521762e-11  3.521762e-11   \n",
       "dDocent_Contig_148-62  ...  213.330700  2.172593e-22  2.172593e-22   \n",
       "dDocent_Contig_152-75  ...  229.337900  9.782448e-25  9.782448e-25   \n",
       "\n",
       "                      P_HET_EXCESS A1_allele  A2_allele   A1_freq   A2_freq  \\\n",
       "dDocent_Contig_23-44      0.635577         T          C  0.956612  0.043388   \n",
       "dDocent_Contig_27-26      0.259856         T          A  0.917553  0.082447   \n",
       "dDocent_Contig_103-30     1.000000         A          C  0.979592  0.020408   \n",
       "dDocent_Contig_148-62     1.000000         C          A  0.934694  0.065306   \n",
       "dDocent_Contig_152-75     1.000000         T          G  0.932653  0.067347   \n",
       "\n",
       "                            MAF       Fis  \n",
       "dDocent_Contig_23-44   0.043388 -0.043197  \n",
       "dDocent_Contig_27-26   0.082447 -0.086957  \n",
       "dDocent_Contig_103-30  0.020408  1.000000  \n",
       "dDocent_Contig_148-62  0.065306  0.933270  \n",
       "dDocent_Contig_152-75  0.067347  0.967575  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_23-44</th>\n",
       "      <td>dDocent_Contig_23</td>\n",
       "      <td>44</td>\n",
       "      <td>2</td>\n",
       "      <td>484</td>\n",
       "      <td>T:0.956612</td>\n",
       "      <td>C:0.0433884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_27-26</th>\n",
       "      <td>dDocent_Contig_27</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>376</td>\n",
       "      <td>T:0.917553</td>\n",
       "      <td>A:0.0824468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_103-30</th>\n",
       "      <td>dDocent_Contig_103</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>490</td>\n",
       "      <td>A:0.979592</td>\n",
       "      <td>C:0.0204082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_148-62</th>\n",
       "      <td>dDocent_Contig_148</td>\n",
       "      <td>62</td>\n",
       "      <td>2</td>\n",
       "      <td>490</td>\n",
       "      <td>C:0.934694</td>\n",
       "      <td>A:0.0653061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_152-75</th>\n",
       "      <td>dDocent_Contig_152</td>\n",
       "      <td>75</td>\n",
       "      <td>2</td>\n",
       "      <td>490</td>\n",
       "      <td>T:0.932653</td>\n",
       "      <td>G:0.0673469</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    CHROM POS N_ALLELES N_CHR     A1_FREQ  \\\n",
       "dDocent_Contig_23-44    dDocent_Contig_23  44         2   484  T:0.956612   \n",
       "dDocent_Contig_27-26    dDocent_Contig_27  26         2   376  T:0.917553   \n",
       "dDocent_Contig_103-30  dDocent_Contig_103  30         2   490  A:0.979592   \n",
       "dDocent_Contig_148-62  dDocent_Contig_148  62         2   490  C:0.934694   \n",
       "dDocent_Contig_152-75  dDocent_Contig_152  75         2   490  T:0.932653   \n",
       "\n",
       "                           A2_FREQ  \n",
       "dDocent_Contig_23-44   C:0.0433884  \n",
       "dDocent_Contig_27-26   A:0.0824468  \n",
       "dDocent_Contig_103-30  C:0.0204082  \n",
       "dDocent_Contig_148-62  A:0.0653061  \n",
       "dDocent_Contig_152-75  G:0.0673469  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_23-44</th>\n",
       "      <td>dDocent_Contig_23</td>\n",
       "      <td>44</td>\n",
       "      <td>221/21/0</td>\n",
       "      <td>221.46/20.09/0.46</td>\n",
       "      <td>0.497843</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.635577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_27-26</th>\n",
       "      <td>dDocent_Contig_27</td>\n",
       "      <td>26</td>\n",
       "      <td>157/31/0</td>\n",
       "      <td>158.28/28.44/1.28</td>\n",
       "      <td>1.517900</td>\n",
       "      <td>6.176162e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.259856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_103-30</th>\n",
       "      <td>dDocent_Contig_103</td>\n",
       "      <td>30</td>\n",
       "      <td>240/0/5</td>\n",
       "      <td>235.10/9.80/0.10</td>\n",
       "      <td>245.000000</td>\n",
       "      <td>3.521762e-11</td>\n",
       "      <td>3.521762e-11</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_148-62</th>\n",
       "      <td>dDocent_Contig_148</td>\n",
       "      <td>62</td>\n",
       "      <td>228/2/15</td>\n",
       "      <td>214.04/29.91/1.04</td>\n",
       "      <td>213.330700</td>\n",
       "      <td>2.172593e-22</td>\n",
       "      <td>2.172593e-22</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_152-75</th>\n",
       "      <td>dDocent_Contig_152</td>\n",
       "      <td>75</td>\n",
       "      <td>228/1/16</td>\n",
       "      <td>213.11/30.78/1.11</td>\n",
       "      <td>229.337900</td>\n",
       "      <td>9.782448e-25</td>\n",
       "      <td>9.782448e-25</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    CHROM  POS OBS(HOM1/HET/HOM2)  \\\n",
       "dDocent_Contig_23-44    dDocent_Contig_23   44           221/21/0   \n",
       "dDocent_Contig_27-26    dDocent_Contig_27   26           157/31/0   \n",
       "dDocent_Contig_103-30  dDocent_Contig_103   30            240/0/5   \n",
       "dDocent_Contig_148-62  dDocent_Contig_148   62           228/2/15   \n",
       "dDocent_Contig_152-75  dDocent_Contig_152   75           228/1/16   \n",
       "\n",
       "                        E(HOM1/HET/HOM2)   ChiSq_HWE         P_HWE  \\\n",
       "dDocent_Contig_23-44   221.46/20.09/0.46    0.497843  1.000000e+00   \n",
       "dDocent_Contig_27-26   158.28/28.44/1.28    1.517900  6.176162e-01   \n",
       "dDocent_Contig_103-30   235.10/9.80/0.10  245.000000  3.521762e-11   \n",
       "dDocent_Contig_148-62  214.04/29.91/1.04  213.330700  2.172593e-22   \n",
       "dDocent_Contig_152-75  213.11/30.78/1.11  229.337900  9.782448e-25   \n",
       "\n",
       "                       P_HET_DEFICIT  P_HET_EXCESS  \n",
       "dDocent_Contig_23-44    1.000000e+00      0.635577  \n",
       "dDocent_Contig_27-26    1.000000e+00      0.259856  \n",
       "dDocent_Contig_103-30   3.521762e-11      1.000000  \n",
       "dDocent_Contig_148-62   2.172593e-22      1.000000  \n",
       "dDocent_Contig_152-75   9.782448e-25      1.000000  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_out = os.path.join(analysis_dir,'loci_out.csv')\n",
    "frq_out = os.path.join(analysis_dir,'frq_out.csv')\n",
    "hardy_out = os.path.join(analysis_dir,'hardy_out.csv')\n",
    "\n",
    "loci_df.to_csv(path_or_buf=loci_out)\n",
    "frq_df.to_csv(path_or_buf=frq_out)\n",
    "hardy.to_csv(path_or_buf=hardy_out)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at the summary stats and make decisions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    34737.000000\n",
       "mean         8.629812\n",
       "std          8.754934\n",
       "min          1.338780\n",
       "25%          4.922450\n",
       "50%          6.632650\n",
       "75%          9.457140\n",
       "max        249.473000\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    34737.000000\n",
       "mean       748.773133\n",
       "std        281.334951\n",
       "min          4.247070\n",
       "25%        496.000000\n",
       "50%        883.000000\n",
       "75%        999.000000\n",
       "max        999.000000\n",
       "Name: QUAL, dtype: float64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.QUAL.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZEAAAD4CAYAAAAtrdtxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAY10lEQVR4nO3dfbRddX3n8ffHREBsJQmkNM1DbywpTqA6xCvE5XRGoQ0BHMPMUBuWLSnNMjM1ttqyRhPrasYHuqDjGGFGGFNIeRiHEJFKitQ0RtDVtUogCOWZcuUpNwZzJQGmYsHgZ/7YvwuH67nJuTv3nJN77+e11lnZ+7t/++zfL5uVD/vh7C3bRERE1PG6bncgIiLGroRIRETUlhCJiIjaEiIREVFbQiQiImqb3O0OdNoxxxzjnp6ebncjImJMueuuu35oe/rQ+oQLkZ6eHrZv397tbkREjCmSnmxWz+msiIioLSESERG1JUQiIqK2hEhERNSWEImIiNoSIhERUVtCJCIiakuIREREbQmRiIiobcL9Yj0iopt6Vn29K9t94qKz2vK9ORKJiIjaEiIREVFbQiQiImpLiERERG0JkYiIqC0hEhERtSVEIiKitoRIRETUlhCJiIja2hYiktZL2i3p/iH1P5T0sKQHJP1FQ321pD5Jj0g6vaG+uNT6JK1qqM+VtK3Ur5d0WLvGEhERzbXzSOQqYHFjQdJ7gCXA22yfAHyu1OcDS4ETyjqXSZokaRLwReAMYD5wbmkLcDGw1vZxwF5geRvHEhERTbQtRGx/B9gzpPwHwEW2Xyxtdpf6EmCD7RdtPw70ASeXT5/tx2y/BGwAlkgScCpwQ1n/auDsdo0lIiKa6/Q1kV8Ffr2chvq2pHeU+kxgR0O7/lIbrn408KztfUPqTUlaIWm7pO0DAwOjNJSIiOh0iEwGpgELgf8KbCxHFW1le53tXtu906dPb/fmIiImjE4/Cr4fuNG2gTsk/RQ4BtgJzG5oN6vUGKb+DDBF0uRyNNLYPiIiOqTTRyJfA94DIOlXgcOAHwKbgKWSDpc0F5gH3AHcCcwrd2IdRnXxfVMJoVuBc8r3LgNu6uhIIiKifUcikq4D3g0cI6kfWAOsB9aX235fApaVQHhA0kbgQWAfsNL2y+V7PgxsBiYB620/UDbxcWCDpM8CdwNXtmssERHRXNtCxPa5wyz6nWHaXwhc2KR+C3BLk/pjVHdvRUREl+QX6xERUVtCJCIiakuIREREbQmRiIioLSESERG1JUQiIqK2hEhERNSWEImIiNoSIhERUVtCJCIiakuIREREbQmRiIioLSESERG1JUQiIqK2hEhERNTWthCRtF7S7vICqqHLLpBkSceUeUm6VFKfpHslLWhou0zSo+WzrKH+dkn3lXUu7cS72iMi4rXaeSRyFbB4aFHSbGAR8FRD+QyqV+LOA1YAl5e206jeiHgK1Quo1kiaWta5HPhgw3o/s62IiGivtoWI7e8Ae5osWgt8DHBDbQlwjSu3A1MkzQBOB7bY3mN7L7AFWFyWvcn27eX1utcAZ7drLBER0VxHr4lIWgLstP2PQxbNBHY0zPeX2v7q/U3qw213haTtkrYPDAwcxAgiIqJRx0JE0pHAJ4A/69Q2B9leZ7vXdu/06dM7vfmIiHGrk0civwLMBf5R0hPALOC7kn4R2AnMbmg7q9T2V5/VpB4RER3UsRCxfZ/tX7DdY7uH6hTUAttPA5uA88pdWguB52zvAjYDiyRNLRfUFwGby7LnJS0sd2WdB9zUqbFERESlnbf4Xgf8A3C8pH5Jy/fT/BbgMaAP+EvgQwC29wCfAe4sn0+XGqXNFWWd7wF/245xRETE8Ca364ttn3uA5T0N0wZWDtNuPbC+SX07cOLB9TIiIg5GfrEeERG1JUQiIqK2hEhERNSWEImIiNoSIhERUVtCJCIiakuIREREbQmRiIioLSESERG1JUQiIqK2hEhERNSWEImIiNoSIhERUVtCJCIiakuIREREbe18KdV6Sbsl3d9Q+++SHpZ0r6S/ljSlYdlqSX2SHpF0ekN9can1SVrVUJ8raVupXy/psHaNJSIimmvnkchVwOIhtS3AibbfCvwTsBpA0nxgKXBCWecySZMkTQK+CJwBzAfOLW0BLgbW2j4O2Avs782JERHRBm0LEdvfAfYMqf2d7X1l9nZgVpleAmyw/aLtx6leeXty+fTZfsz2S8AGYEl5r/qpwA1l/auBs9s1loiIaK6b10R+n1ffiz4T2NGwrL/UhqsfDTzbEEiD9YiI6KCuhIikPwX2AV/u0PZWSNouafvAwEAnNhkRMSF0PEQk/R7wXuADtl3KO4HZDc1mldpw9WeAKZImD6k3ZXud7V7bvdOnTx+VcURERIdDRNJi4GPA+2y/0LBoE7BU0uGS5gLzgDuAO4F55U6sw6guvm8q4XMrcE5ZfxlwU6fGERERlXbe4nsd8A/A8ZL6JS0H/hfw88AWSfdI+t8Ath8ANgIPAt8AVtp+uVzz+DCwGXgI2FjaAnwc+BNJfVTXSK5s11giIqK5yQduUo/tc5uUh/2H3vaFwIVN6rcAtzSpP0Z191ZERHRJfrEeERG1JUQiIqK2hEhERNSWEImIiNoSIhERUVtCJCIiakuIREREbQmRiIioraUQkfRr7e5IRESMPa0eiVwm6Q5JH5J0VFt7FBERY0ZLIWL714EPUD1R9y5J/1fSb7a1ZxERcchr+ZqI7UeBT1I9+PDfAZeW96X/x3Z1LiIiDm2tXhN5q6S1VE/SPRX497b/VZle28b+RUTEIazVp/j+T+AK4BO2fzxYtP19SZ9sS88iIuKQ12qInAX82PbLAJJeBxxh+wXb17atdxERcUhr9ZrIN4E3NMwfWWoRETGBtRoiR9j+58GZMn3k/laQtF7Sbkn3N9SmSdoi6dHy59RSl6RLJfVJulfSgoZ1lpX2j0pa1lB/u6T7yjqXSlKrg46IiNHRaoj8aMg/7G8Hfryf9gBXAYuH1FYBW23PA7aWeYAzqN6rPg9YAVxetjMNWAOcQvUWwzWDwVPafLBhvaHbioiINmv1mshHga9I+j4g4BeB397fCra/I6lnSHkJ8O4yfTVwG9Utw0uAa2wbuF3SFEkzStsttvcASNoCLJZ0G/Am27eX+jXA2cDftjieiIgYBS2FiO07Jb0FOL6UHrH9kxrbO9b2rjL9NHBsmZ4J7Gho119q+6v3N6k3JWkF1REOc+bMqdHtiIhoptUjEYB3AD1lnQWSsH1N3Q3btiTXXX+E21oHrAPo7e3tyDYjIiaClkJE0rXArwD3AC+XsoGRhsgPJM2wvaucrtpd6jupHqkyaFap7eTV01+D9dtKfVaT9hER0UGtHon0AvPLNYuDsQlYBlxU/rypof5hSRuoLqI/V4JmM/DnDRfTFwGrbe+R9LykhcA24DyqH0RGREQHtRoi91NdTN91oIaDJF1HdRRxjKR+qrusLgI2SloOPAm8vzS/BTgT6ANeAM4HKGHxGeDO0u7TgxfZgQ9R3QH2BqoL6rmoHhHRYa2GyDHAg5LuAF4cLNp+33Ar2D53mEWnNWlrYOUw37MeWN+kvh04cf/djoiIdmo1RP5bOzsRERFjU6u3+H5b0i8D82x/U9KRwKT2di0iIg51rT4K/oPADcCXSmkm8LV2dSoiIsaGVh97shJ4F/A8vPKCql9oV6ciImJsaDVEXrT90uCMpMlUvxOJiIgJrNUQ+bakTwBvKO9W/wrwN+3rVkREjAWthsgqYAC4D/jPVL/ryBsNIyImuFbvzvop8JflExERAbT+7KzHaXINxPabR71HERExZozk2VmDjgB+C5g2+t2JiIixpKVrIrafafjstP0F4Kw29y0iIg5xrZ7OWtAw+zqqI5ORvIskIiLGoVaD4H80TO8DnuDVJ/BGRMQE1erdWe9pd0ciImLsafV01p/sb7ntz49OdyIiYixp9ceGvcAfUD14cSbwX4AFwM+Xz4hI+mNJD0i6X9J1ko6QNFfSNkl9kq6XdFhpe3iZ7yvLexq+Z3WpPyLp9JH2IyIiDk6rITILWGD7AtsXAG8H5tj+lO1PjWSDkmYCfwT02j6R6pHyS4GLgbW2jwP2AsvLKsuBvaW+trRD0vyy3gnAYuAySXk8fUREB7UaIscCLzXMv1RqdU2meg7XZOBIqtfunkr1uHmAq4Gzy/SSMk9ZfpoklfoG2y/afpzq1bonH0SfIiJihFq9O+sa4A5Jf13mz+bVf9hHxPZOSZ8DngJ+DPwdcBfwrO19pVk/1Wkzyp87yrr7JD0HHF3qtzd8deM6ryFpBbACYM6cOXW6HRERTbT6Y8MLgfOpTjPtBc63/ed1NihpKtVRxFzgl4A3Up2Oahvb62z32u6dPn16OzcVETGhtHo6C6rTTs/bvgTolzS35jZ/A3jc9oDtnwA3Ur3wako5vQXVNZidZXonMBteeY/JUcAzjfUm60RERAe0+nrcNcDHgdWl9Hrg/9Tc5lPAQklHlmsbpwEPArcC55Q2y4CbyvSmMk9Z/i3bLvWl5e6tucA84I6afYqIiBpavSbyH4CTgO8C2P6+pBHf2lvW3SbphvJd+4C7gXXA14ENkj5baleWVa4ErpXUB+yhuiML2w9I2kgVQPuAlbZfrtOniIiop9UQecm2JRlA0hsPZqO21wBrhpQfo8ndVbb/heqpwc2+50LgwoPpS0RE1NfqNZGNkr5Edd3ig8A3yQuqIiImvAMeiZTrFtcDbwGeB44H/sz2ljb3LSIiDnEHDJFyGusW278GJDgiIuIVrZ7O+q6kd7S1JxERMea0emH9FOB3JD0B/AgQ1UHKW9vVsYiIOPTtN0QkzbH9FJAn5EZExM840JHI16ie3vukpK/a/k+d6FRERIwNB7omoobpN7ezIxERMfYcKEQ8zHRERMQBT2e9TdLzVEckbyjT8OqF9Te1tXcREXFI22+I2M6bAiMiYlgjeRR8RETEayREIiKitoRIRETUlhCJiIjauhIikqZIukHSw5IekvROSdMkbZH0aPlzamkrSZdK6pN0r6QFDd+zrLR/VNKy4bcYERHt0K0jkUuAb9h+C/A24CFgFbDV9jxga5kHOIPq1bfzgBXA5QCSplG92OoUqpdZrRkMnoiI6IyOh4iko4B/S3n9re2XbD8LLAGuLs2uBs4u00uAa1y5nerFWDOonue1xfYe23upHlO/uINDiYiY8LpxJDIXGAD+StLdkq4or9s91vau0uZp4NgyPRPY0bB+f6kNV/8ZklZI2i5p+8DAwCgOJSJiYutGiEwGFgCX2z6J6tHyqxob2Daj+JgV2+ts99runT59+mh9bUTEhNeNEOkH+m1vK/M3UIXKD8ppKsqfu8vyncDshvVnldpw9YiI6JCOh4jtp4Edko4vpdOAB4FNwOAdVsuAm8r0JuC8cpfWQuC5ctprM7BI0tRyQX1RqUVERIe0+mbD0faHwJclHQY8BpxPFWgbJS0HngTeX9reApwJ9AEvlLbY3iPpM8Cdpd2nbe/p3BAiIqIrIWL7HqC3yaLTmrQ1sHKY71kPrB/d3kVERKvyi/WIiKgtIRIREbUlRCIioraESERE1JYQiYiI2hIiERFRW0IkIiJqS4hERERtCZGIiKgtIRIREbUlRCIioraESERE1JYQiYiI2hIiERFRW0IkIiJq61qISJok6W5JN5f5uZK2SeqTdH15YRWSDi/zfWV5T8N3rC71RySd3p2RRERMXN08EvkI8FDD/MXAWtvHAXuB5aW+HNhb6mtLOyTNB5YCJwCLgcskTepQ3yMigi6FiKRZwFnAFWVewKnADaXJ1cDZZXpJmacsP620XwJssP2i7cepXp97cmdGEBER0L0jkS8AHwN+WuaPBp61va/M9wMzy/RMYAdAWf5caf9Kvck6ryFphaTtkrYPDAyM5jgiIia0joeIpPcCu23f1alt2l5nu9d27/Tp0zu12YiIcW9yF7b5LuB9ks4EjgDeBFwCTJE0uRxtzAJ2lvY7gdlAv6TJwFHAMw31QY3rREREB3T8SMT2atuzbPdQXRj/lu0PALcC55Rmy4CbyvSmMk9Z/i3bLvWl5e6tucA84I4ODSMiIujOkchwPg5skPRZ4G7gylK/ErhWUh+whyp4sP2ApI3Ag8A+YKXtlzvf7YiIiaurIWL7NuC2Mv0YTe6usv0vwG8Ns/6FwIXt62FEROxPfrEeERG1JUQiIqK2hEhERNSWEImIiNoSIhERUduhdItvxCGhZ9XXu7LdJy46qyvbjTgYORKJiIjaEiIREVFbTmdFHCK6dRoNciot6kuIxCGpm/+gRkTrcjorIiJqS4hERERtCZGIiKgtIRIREbUlRCIioraO350laTZwDXAsYGCd7UskTQOuB3qAJ4D3294rSVSvzz0TeAH4PdvfLd+1DPhk+erP2r66k2OJGC/yK/2oqxtHIvuAC2zPBxYCKyXNB1YBW23PA7aWeYAzqF59Ow9YAVwOUEJnDXAK1cus1kia2smBRERMdN14x/quwSMJ2/8PeAiYCSwBBo8krgbOLtNLgGtcuR2YImkGcDqwxfYe23uBLcDiDg4lImLC6+o1EUk9wEnANuBY27vKoqepTndBFTA7GlbrL7Xh6hER0SFd+8W6pJ8Dvgp81Pbz1aWPim1L8ihuawXVqTDmzJkzWl8bEQcp12LGvq6EiKTXUwXIl23fWMo/kDTD9q5yump3qe8EZjesPqvUdgLvHlK/rdn2bK8D1gH09vaOWjhNBHn8SETsT8dPZ5W7ra4EHrL9+YZFm4BlZXoZcFND/TxVFgLPldNem4FFkqaWC+qLSi0iIjqkG0ci7wJ+F7hP0j2l9gngImCjpOXAk8D7y7JbqG7v7aO6xfd8ANt7JH0GuLO0+7TtPZ0ZQkREQBdCxPbfAxpm8WlN2htYOcx3rQfWj17vIiJiJPIo+IiYcHKtb/TksScREVFbQiQiImpLiERERG0JkYiIqC0hEhERtSVEIiKitoRIRETUlhCJiIja8mPDMSA/jIqIQ1WORCIioraESERE1JYQiYiI2hIiERFRW0IkIiJqS4hERERtY/4WX0mLgUuAScAVti9q17Zyq21ExGuN6SMRSZOALwJnAPOBcyXN726vIiImjjEdIsDJQJ/tx2y/BGwAlnS5TxERE8ZYP501E9jRMN8PnDK0kaQVwIoy+8+SHqmxrWOAH9ZYbyzLmCeGjHkC0MXAwY37l5sVx3qItMT2OmDdwXyHpO22e0epS2NCxjwxZMwTRzvGPdZPZ+0EZjfMzyq1iIjogLEeIncC8yTNlXQYsBTY1OU+RURMGGP6dJbtfZI+DGymusV3ve0H2rS5gzodNkZlzBNDxjxxjPq4ZXu0vzMiIiaIsX46KyIiuighEhERtSVEDkDSYkmPSOqTtKrb/RktkmZLulXSg5IekPSRUp8maYukR8ufU0tdki4tfw/3SlrQ3RHUJ2mSpLsl3Vzm50raVsZ2fblJA0mHl/m+srynm/0+GJKmSLpB0sOSHpL0zvG+ryX9cflv+35J10k6Yrzta0nrJe2WdH9DbcT7VdKy0v5RSctG0oeEyH6M88eq7AMusD0fWAisLGNbBWy1PQ/YWuah+juYVz4rgMs73+VR8xHgoYb5i4G1to8D9gLLS305sLfU15Z2Y9UlwDdsvwV4G9X4x+2+ljQT+COg1/aJVDfeLGX87eurgMVDaiPar5KmAWuofqh9MrBmMHhaYjufYT7AO4HNDfOrgdXd7lebxnoT8JvAI8CMUpsBPFKmvwSc29D+lXZj6UP1W6KtwKnAzYCofsE7eeg+p7rr751lenJpp26PocaYjwIeH9r38byvefVpFtPKvrsZOH087mugB7i/7n4FzgW+1FB/TbsDfXIksn/NHqsys0t9aZty6H4SsA041vausuhp4NgyPV7+Lr4AfAz4aZk/GnjW9r4y3ziuV8Zclj9X2o81c4EB4K/KabwrJL2Rcbyvbe8EPgc8Beyi2nd3Mf73NYx8vx7U/k6ITHCSfg74KvBR2883LnP1vyXj5h5wSe8Fdtu+q9t96bDJwALgctsnAT/i1VMcwLjc11OpHsY6F/gl4I387Gmfca8T+zUhsn/j+rEqkl5PFSBftn1jKf9A0oyyfAawu9THw9/Fu4D3SXqC6onPp1JdK5giafCHt43jemXMZflRwDOd7PAo6Qf6bW8r8zdQhcp43te/ATxue8D2T4Abqfb/eN/XMPL9elD7OyGyf+P2sSqSBFwJPGT78w2LNgGDd2cso7pWMlg/r9zhsRB4ruGQeUywvdr2LNs9VPvyW7Y/ANwKnFOaDR3z4N/FOaX9mPu/ddtPAzskHV9KpwEPMo73NdVprIWSjiz/rQ+OeVzv62Kk+3UzsEjS1HIEt6jUWtPti0KH+gc4E/gn4HvAn3a7P6M4rn9DdZh7L3BP+ZxJdR54K/Ao8E1gWmkvqjvVvgfcR3XXS9fHcRDjfzdwc5l+M3AH0Ad8BTi81I8o831l+Zu73e+DGO+/BraX/f01YOp439fAp4CHgfuBa4HDx9u+Bq6juubzE6ojzuV19ivw+2XsfcD5I+lDHnsSERG15XRWRETUlhCJiIjaEiIREVFbQiQiImpLiERERG0JkYiIqC0hEhERtf1/N9BijtpWzi4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this would not be zero if there was an error in the calculation\n",
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(981, 34305)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.MEAN_DEPTH > 25]),len(loci_df[loci_df.MEAN_DEPTH >= 3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "760"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 100]) - len(loci_df[loci_df.QUAL >= 200])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(943, 1753, 8788)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL <  200]), len(loci_df[loci_df.QUAL < 250]), len(loci_df[loci_df.QUAL < 500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92, 87)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " len(loci_df[loci_df.Fis <= -0.5]), len(loci_df[loci_df.MAF < 0.02])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) &  \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.MEAN_DEPTH >= 3) & \n",
    "                  (df.MEAN_DEPTH < 25) & \n",
    "                  (df.QUAL >= 250) & \n",
    "                  (df.MAF >= 0.02) &  \n",
    "                  (df.Fis > -0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31652, 25)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/good_snps\n",
      "\t--positions /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 31652 out of a possible 34737 Sites\n",
      "Run Time = 18.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_bad_remove_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"stage1_positions.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "snps_gz = snps + \".gz\"\n",
    "!$bgzip -c {snps} > {snps_gz}\n",
    "!$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make 012, see directory PCA_012 for results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/good_snps.recode.vcf.gz\n",
      "\t--012\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/good_snps.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 31652 out of a possible 31652 Sites\n",
      "Run Time = 3.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf {snps_gz} \\\n",
    "--out {snps_gz} \\\n",
    "--012"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get coverage per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/good_snps.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall/filtering/good_snps.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 31652 out of a possible 31652 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf {snps_gz} \\\n",
    "--out {snps_gz} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PH_AS_10</td>\n",
       "      <td>31652</td>\n",
       "      <td>9.84706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PH_AS_11</td>\n",
       "      <td>31652</td>\n",
       "      <td>5.68890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PH_AS_12</td>\n",
       "      <td>31652</td>\n",
       "      <td>10.21510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PH_AS_1</td>\n",
       "      <td>31652</td>\n",
       "      <td>10.24440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PH_AS_2</td>\n",
       "      <td>31652</td>\n",
       "      <td>9.31246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       INDV  N_SITES  MEAN_DEPTH\n",
       "0  PH_AS_10    31652     9.84706\n",
       "1  PH_AS_11    31652     5.68890\n",
       "2  PH_AS_12    31652    10.21510\n",
       "3   PH_AS_1    31652    10.24440\n",
       "4   PH_AS_2    31652     9.31246"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = os.path.join(analysis_dir, \"good_snps.recode.vcf.gz.idepth\")\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    245.000000\n",
       "mean       7.739640\n",
       "std        2.023261\n",
       "min        2.362980\n",
       "25%        6.302410\n",
       "50%        7.780460\n",
       "75%        9.216040\n",
       "max       17.353300\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimate relatedness\n",
    "\n",
    "#### Papers:\n",
    "KING: 10.1093/bioinformatics/btq559  \n",
    "Error in KING: 10.3389/fgene.2022.882268  \n",
    "\n",
    "#### Relationship from papers:   \n",
    "Duplicate/Clone/twin: r > 0.354    \n",
    "Parent-offspring/full-sib: 0.177 < r < 0.354    \n",
    "half-sib/cousin/2nd-degree: 0.0884 < r < 0.177  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/SNPcall_tf/filtering/good_snps.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/SNPcall_tf/filtering/good_snps.recode.vcf.gz\n",
      "\t--relatedness2\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 245 out of 245 Individuals\n",
      "Outputting Individual Relatedness\n",
      "After filtering, kept 31652 out of a possible 31652 Sites\n",
      "Run Time = 13.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf {snps_gz} \\\n",
    "--out {snps_gz} \\\n",
    "--relatedness2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove unnecessary files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm snps*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *miss*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check 012 that PCA looks okay  \n",
    "\n",
    "scp ggood_snps.recode.vcf.gz, good_snps.recode.vcf.gz.012, good_snps.recode.vcf.gz.012.indv over to your local computer and run same R markdown as in the ddocent_output, PCA_012 dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
