{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/bin/python\r\n"
     ]
    }
   ],
   "source": [
    "!which python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/bin/sh: line 0: source: activate: file not found\r\n"
     ]
    }
   ],
   "source": [
    "!source activate py36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "#sys.path.append('/home/faske/g/anaconda3/envs/py34/lib/python3.4/site-packages')\n",
    "sys.path.append('/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages')\n",
    "sys.path.append(\"/data/gpfs/assoc/parchmanlab/tfaske/ipynb/include_utils\")\n",
    "\n",
    "import ipyparallel as ipp\n",
    "import os, time\n",
    "import include_utils as u\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import random\n",
    "import numbers\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "#import vcf\n",
    "from sklearn import preprocessing\n",
    "from subprocess import Popen, PIPE, call, check_output\n",
    "import seaborn as sns\n",
    "from IPython.display import FileLink\n",
    "import urllib.request as urllib2\n",
    "import dill\n",
    "import traceback\n",
    "from pandas import Series, DataFrame\n",
    "import gzip\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore',category=pd.io.pytables.PerformanceWarning)\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "from Bio import SeqIO\n",
    "#import pysam\n",
    "from collections import OrderedDict, namedtuple, Counter\n",
    "import operator\n",
    "import multiprocessing as mp\n",
    "import shutil\n",
    "import tempfile\n",
    "#from ipyparallel import Client\n",
    "import scandir\n",
    "import glob\n",
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "import pickle\n",
    "import re\n",
    "from itertools import chain\n",
    "#import Levenshtein as lv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfutils = \"vcfutils.pl\"\n",
    "vcftools = \"vcftools\"\n",
    "bcftools = \"bcftools\"\n",
    "bgzip = \"bgzip\"\n",
    "tabix = \"tabix\"\n",
    "java = \"java\"\n",
    "plink = \"plink --noweb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = '/data/gpfs/assoc/denovo/PHHA/ddocent/filtering/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/PHHA/ddocent/filtering\n"
     ]
    }
   ],
   "source": [
    "cd $analysis_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/PHHA/ddocent/filtering/TotalRawSNPs.vcf.gz'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file = os.path.join(analysis_dir, \"TotalRawSNPs.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keep only biallelic first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/TotalRawSNPs.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--out PHHA.biallelic\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 256 out of 256 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 775804 out of a possible 1113694 Sites\n",
      "Run Time = 646.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $'PHHA.biallelic'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_biallelic = os.path.join(analysis_dir, \"PHHA.biallelic.recode.vcf\")\n",
    "vcf_biallelic_gz = vcf_biallelic + '.gz'\n",
    "!$bgzip -c $vcf_biallelic > {vcf_biallelic_gz}\n",
    "!$tabix {vcf_biallelic_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove by MAF, missing, and thin next"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/PHHA.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--thin 100\n",
      "\t--max-missing 0.6\n",
      "\t--out PHHA_miss60_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 256 out of 256 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 33703 out of a possible 775804 Sites\n",
      "Run Time = 68.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.6 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'PHHA_miss60_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/PHHA.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--thin 100\n",
      "\t--max-missing 0.7\n",
      "\t--out PHHA_miss70_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 256 out of 256 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 28114 out of a possible 775804 Sites\n",
      "Run Time = 63.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.7 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'PHHA_miss70_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"PHHA_miss70_thin100_MAF2.recode.vcf\"\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove bad_indv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf PHHA_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--missing-indv\n",
      "\t--out PHHA_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 256 out of 256 Individuals\n",
      "Outputting Individual Missingness\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz --out $vcf_filtered_gz --missing-indv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_imiss(vcf_file):\n",
    "    imiss_file = !ls {vcf_file}.imiss\n",
    "    imiss_df = pd.read_csv(imiss_file[0], sep=\"\\t\")\n",
    "    imiss_df.index = imiss_df.INDV\n",
    "    return imiss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPES_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INDV</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PH_AS_10</th>\n",
       "      <td>PH_AS_10</td>\n",
       "      <td>28114</td>\n",
       "      <td>0</td>\n",
       "      <td>953</td>\n",
       "      <td>0.033898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_11</th>\n",
       "      <td>PH_AS_11</td>\n",
       "      <td>28114</td>\n",
       "      <td>0</td>\n",
       "      <td>2671</td>\n",
       "      <td>0.095006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_12</th>\n",
       "      <td>PH_AS_12</td>\n",
       "      <td>28114</td>\n",
       "      <td>0</td>\n",
       "      <td>1788</td>\n",
       "      <td>0.063598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_1</th>\n",
       "      <td>PH_AS_1</td>\n",
       "      <td>28114</td>\n",
       "      <td>0</td>\n",
       "      <td>1842</td>\n",
       "      <td>0.065519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PH_AS_2</th>\n",
       "      <td>PH_AS_2</td>\n",
       "      <td>28114</td>\n",
       "      <td>0</td>\n",
       "      <td>590</td>\n",
       "      <td>0.020986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              INDV  N_DATA  N_GENOTYPES_FILTERED  N_MISS    F_MISS\n",
       "INDV                                                              \n",
       "PH_AS_10  PH_AS_10   28114                     0     953  0.033898\n",
       "PH_AS_11  PH_AS_11   28114                     0    2671  0.095006\n",
       "PH_AS_12  PH_AS_12   28114                     0    1788  0.063598\n",
       "PH_AS_1    PH_AS_1   28114                     0    1842  0.065519\n",
       "PH_AS_2    PH_AS_2   28114                     0     590  0.020986"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df = get_imiss(vcf_filtered_gz)\n",
    "imiss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    256.000000\n",
       "mean       0.121079\n",
       "std        0.139689\n",
       "min        0.012165\n",
       "25%        0.063171\n",
       "50%        0.080476\n",
       "75%        0.132114\n",
       "max        0.996230\n",
       "Name: F_MISS, dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df.F_MISS.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(256, 7, 10, 12)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(imiss_df),len(imiss_df[imiss_df.F_MISS >= .5]),len(imiss_df[imiss_df.F_MISS >= .4]), len(imiss_df[imiss_df.F_MISS >= .25])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_indv = imiss_df.INDV[imiss_df.F_MISS >= .25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"bad_indv.txt\"), \"w\") as o:\n",
    "    o.write(\"INDV\\n\")\n",
    "    for elem in bad_indv.index:\n",
    "        o.write(\"%s\\n\" % elem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf PHHA_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--remove /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/bad_indv.txt\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "Excluding individuals in 'exclude' list\n",
      "After filtering, kept 244 out of 256 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 23.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--remove {os.path.join(analysis_dir, \"bad_indv.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"snps_indv_removed\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_bad_remove = os.path.join(analysis_dir, \"snps_indv_removed.recode.vcf\")\n",
    "vcf_bad_remove_gz = vcf_bad_remove + \".gz\"\n",
    "!$bgzip -c {vcf_bad_remove} > {vcf_bad_remove_gz}\n",
    "!$tabix {vcf_bad_remove_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter snps further \n",
    "this needs to be done after removing individuals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Depth for Each Site\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Quality for Each Site\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Frequency Statistics...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 28114 out of a possible 28114 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_bad_remove_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    \n",
    "    hardy_files = !ls {filedir}/{prefix}.hwe\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_files = !ls {filedir}/{prefix}.l* | grep -v log\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    frq_files = !ls {filedir}/{prefix}.frq* | grep -v count\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_bad_remove_gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:45: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/pandas/core/indexing.py:822: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=6, step=1)\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(analysis_dir,'snps_indv_removed.recode.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>...</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_123-64</th>\n",
       "      <td>341959</td>\n",
       "      <td>1726515051</td>\n",
       "      <td>1401.47000</td>\n",
       "      <td>5.132800e+06</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6458.970</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>18.787750</td>\n",
       "      <td>6.680879e-07</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.970073e-07</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.782787</td>\n",
       "      <td>0.217213</td>\n",
       "      <td>0.217213</td>\n",
       "      <td>-0.274869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_237-19</th>\n",
       "      <td>417634</td>\n",
       "      <td>917890644</td>\n",
       "      <td>1711.61000</td>\n",
       "      <td>8.356470e+05</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>814738.000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>186.190600</td>\n",
       "      <td>2.141909e-20</td>\n",
       "      <td>2.141909e-20</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.930328</td>\n",
       "      <td>0.069672</td>\n",
       "      <td>0.069672</td>\n",
       "      <td>0.873802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_278-64</th>\n",
       "      <td>435505</td>\n",
       "      <td>1112584829</td>\n",
       "      <td>1784.86000</td>\n",
       "      <td>1.379720e+06</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>289669.000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>89.268270</td>\n",
       "      <td>2.421640e-06</td>\n",
       "      <td>2.421640e-06</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.973361</td>\n",
       "      <td>0.026639</td>\n",
       "      <td>0.026639</td>\n",
       "      <td>0.605668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_326-42</th>\n",
       "      <td>380410</td>\n",
       "      <td>1123095298</td>\n",
       "      <td>1559.06000</td>\n",
       "      <td>2.181130e+06</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26641.100</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.317867</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>7.493871e-01</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.965164</td>\n",
       "      <td>0.034836</td>\n",
       "      <td>0.034836</td>\n",
       "      <td>-0.033970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_330-22</th>\n",
       "      <td>362256</td>\n",
       "      <td>719827618</td>\n",
       "      <td>1484.66000</td>\n",
       "      <td>7.489800e+05</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>641078.000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>145.672900</td>\n",
       "      <td>1.730150e-14</td>\n",
       "      <td>1.730150e-14</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.942623</td>\n",
       "      <td>0.057377</td>\n",
       "      <td>0.057377</td>\n",
       "      <td>0.773137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441023-46</th>\n",
       "      <td>1728</td>\n",
       "      <td>23932</td>\n",
       "      <td>8.26794</td>\n",
       "      <td>4.637020e+01</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>0.143443</td>\n",
       "      <td>538.412</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.152666</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.737299e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.973684</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>-0.024570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441052-5</th>\n",
       "      <td>1484</td>\n",
       "      <td>19126</td>\n",
       "      <td>7.64948</td>\n",
       "      <td>4.028060e+01</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.204918</td>\n",
       "      <td>708.045</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.197601</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.391836e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.969072</td>\n",
       "      <td>0.030928</td>\n",
       "      <td>0.030928</td>\n",
       "      <td>-0.029255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441078-18</th>\n",
       "      <td>728</td>\n",
       "      <td>3884</td>\n",
       "      <td>3.65829</td>\n",
       "      <td>6.165470e+00</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>0.184426</td>\n",
       "      <td>335.542</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>64.623930</td>\n",
       "      <td>1.872976e-06</td>\n",
       "      <td>1.872976e-06</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.957286</td>\n",
       "      <td>0.042714</td>\n",
       "      <td>0.042714</td>\n",
       "      <td>0.570943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441136-5</th>\n",
       "      <td>857</td>\n",
       "      <td>6041</td>\n",
       "      <td>4.65761</td>\n",
       "      <td>1.119910e+01</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>120</td>\n",
       "      <td>0.245902</td>\n",
       "      <td>685.675</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>113.507600</td>\n",
       "      <td>9.809512e-13</td>\n",
       "      <td>9.809512e-13</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.932065</td>\n",
       "      <td>0.067935</td>\n",
       "      <td>0.067935</td>\n",
       "      <td>0.786006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441160-23</th>\n",
       "      <td>1294</td>\n",
       "      <td>12668</td>\n",
       "      <td>6.04673</td>\n",
       "      <td>2.273960e+01</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>0.122951</td>\n",
       "      <td>440.286</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>48.383460</td>\n",
       "      <td>1.176736e-05</td>\n",
       "      <td>1.176736e-05</td>\n",
       "      <td>9.999998e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.953271</td>\n",
       "      <td>0.046729</td>\n",
       "      <td>0.046729</td>\n",
       "      <td>0.476716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28114 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH     VAR_DEPTH  \\\n",
       "dDocent_Contig_123-64        341959   1726515051  1401.47000  5.132800e+06   \n",
       "dDocent_Contig_237-19        417634    917890644  1711.61000  8.356470e+05   \n",
       "dDocent_Contig_278-64        435505   1112584829  1784.86000  1.379720e+06   \n",
       "dDocent_Contig_326-42        380410   1123095298  1559.06000  2.181130e+06   \n",
       "dDocent_Contig_330-22        362256    719827618  1484.66000  7.489800e+05   \n",
       "...                             ...          ...         ...           ...   \n",
       "dDocent_Contig_441023-46       1728        23932     8.26794  4.637020e+01   \n",
       "dDocent_Contig_441052-5        1484        19126     7.64948  4.028060e+01   \n",
       "dDocent_Contig_441078-18        728         3884     3.65829  6.165470e+00   \n",
       "dDocent_Contig_441136-5         857         6041     4.65761  1.119910e+01   \n",
       "dDocent_Contig_441160-23       1294        12668     6.04673  2.273960e+01   \n",
       "\n",
       "                          N_DATA  N_GENOTYPE_FILTERED  N_MISS    F_MISS  \\\n",
       "dDocent_Contig_123-64        488                    0       0  0.000000   \n",
       "dDocent_Contig_237-19        488                    0       0  0.000000   \n",
       "dDocent_Contig_278-64        488                    0       0  0.000000   \n",
       "dDocent_Contig_326-42        488                    0       0  0.000000   \n",
       "dDocent_Contig_330-22        488                    0       0  0.000000   \n",
       "...                          ...                  ...     ...       ...   \n",
       "dDocent_Contig_441023-46     488                    0      70  0.143443   \n",
       "dDocent_Contig_441052-5      488                    0     100  0.204918   \n",
       "dDocent_Contig_441078-18     488                    0      90  0.184426   \n",
       "dDocent_Contig_441136-5      488                    0     120  0.245902   \n",
       "dDocent_Contig_441160-23     488                    0      60  0.122951   \n",
       "\n",
       "                                QUAL N_ALLELES  ...   ChiSq_HWE         P_HWE  \\\n",
       "dDocent_Contig_123-64       6458.970         2  ...   18.787750  6.680879e-07   \n",
       "dDocent_Contig_237-19     814738.000         2  ...  186.190600  2.141909e-20   \n",
       "dDocent_Contig_278-64     289669.000         2  ...   89.268270  2.421640e-06   \n",
       "dDocent_Contig_326-42      26641.100         2  ...    0.317867  1.000000e+00   \n",
       "dDocent_Contig_330-22     641078.000         2  ...  145.672900  1.730150e-14   \n",
       "...                              ...       ...  ...         ...           ...   \n",
       "dDocent_Contig_441023-46     538.412         2  ...    0.152666  1.000000e+00   \n",
       "dDocent_Contig_441052-5      708.045         2  ...    0.197601  1.000000e+00   \n",
       "dDocent_Contig_441078-18     335.542         2  ...   64.623930  1.872976e-06   \n",
       "dDocent_Contig_441136-5      685.675         2  ...  113.507600  9.809512e-13   \n",
       "dDocent_Contig_441160-23     440.286         2  ...   48.383460  1.176736e-05   \n",
       "\n",
       "                         P_HET_DEFICIT  P_HET_EXCESS A1_allele  A2_allele  \\\n",
       "dDocent_Contig_123-64     1.000000e+00  3.970073e-07         T          G   \n",
       "dDocent_Contig_237-19     2.141909e-20  1.000000e+00         A          C   \n",
       "dDocent_Contig_278-64     2.421640e-06  1.000000e+00         T          C   \n",
       "dDocent_Contig_326-42     1.000000e+00  7.493871e-01         C          T   \n",
       "dDocent_Contig_330-22     1.730150e-14  1.000000e+00         C          A   \n",
       "...                                ...           ...       ...        ...   \n",
       "dDocent_Contig_441023-46  1.000000e+00  8.737299e-01         G          T   \n",
       "dDocent_Contig_441052-5   1.000000e+00  8.391836e-01         A          T   \n",
       "dDocent_Contig_441078-18  1.872976e-06  1.000000e+00         C          T   \n",
       "dDocent_Contig_441136-5   9.809512e-13  1.000000e+00         A          C   \n",
       "dDocent_Contig_441160-23  1.176736e-05  9.999998e-01         A          G   \n",
       "\n",
       "                           A1_freq   A2_freq       MAF       Fis  \n",
       "dDocent_Contig_123-64     0.782787  0.217213  0.217213 -0.274869  \n",
       "dDocent_Contig_237-19     0.930328  0.069672  0.069672  0.873802  \n",
       "dDocent_Contig_278-64     0.973361  0.026639  0.026639  0.605668  \n",
       "dDocent_Contig_326-42     0.965164  0.034836  0.034836 -0.033970  \n",
       "dDocent_Contig_330-22     0.942623  0.057377  0.057377  0.773137  \n",
       "...                            ...       ...       ...       ...  \n",
       "dDocent_Contig_441023-46  0.973684  0.026316  0.026316 -0.024570  \n",
       "dDocent_Contig_441052-5   0.969072  0.030928  0.030928 -0.029255  \n",
       "dDocent_Contig_441078-18  0.957286  0.042714  0.042714  0.570943  \n",
       "dDocent_Contig_441136-5   0.932065  0.067935  0.067935  0.786006  \n",
       "dDocent_Contig_441160-23  0.953271  0.046729  0.046729  0.476716  \n",
       "\n",
       "[28114 rows x 25 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_123-64</th>\n",
       "      <td>dDocent_Contig_123</td>\n",
       "      <td>64</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.782787</td>\n",
       "      <td>G:0.217213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_237-19</th>\n",
       "      <td>dDocent_Contig_237</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>A:0.930328</td>\n",
       "      <td>C:0.0696721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_278-64</th>\n",
       "      <td>dDocent_Contig_278</td>\n",
       "      <td>64</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>T:0.973361</td>\n",
       "      <td>C:0.0266393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_326-42</th>\n",
       "      <td>dDocent_Contig_326</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.965164</td>\n",
       "      <td>T:0.0348361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_330-22</th>\n",
       "      <td>dDocent_Contig_330</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>C:0.942623</td>\n",
       "      <td>A:0.057377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441023-46</th>\n",
       "      <td>dDocent_Contig_441023</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>418</td>\n",
       "      <td>G:0.973684</td>\n",
       "      <td>T:0.0263158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441052-5</th>\n",
       "      <td>dDocent_Contig_441052</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>388</td>\n",
       "      <td>A:0.969072</td>\n",
       "      <td>T:0.0309278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441078-18</th>\n",
       "      <td>dDocent_Contig_441078</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>398</td>\n",
       "      <td>C:0.957286</td>\n",
       "      <td>T:0.0427136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441136-5</th>\n",
       "      <td>dDocent_Contig_441136</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>368</td>\n",
       "      <td>A:0.932065</td>\n",
       "      <td>C:0.0679348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441160-23</th>\n",
       "      <td>dDocent_Contig_441160</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>428</td>\n",
       "      <td>A:0.953271</td>\n",
       "      <td>G:0.046729</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28114 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CHROM POS N_ALLELES N_CHR  \\\n",
       "dDocent_Contig_123-64        dDocent_Contig_123  64         2   488   \n",
       "dDocent_Contig_237-19        dDocent_Contig_237  19         2   488   \n",
       "dDocent_Contig_278-64        dDocent_Contig_278  64         2   488   \n",
       "dDocent_Contig_326-42        dDocent_Contig_326  42         2   488   \n",
       "dDocent_Contig_330-22        dDocent_Contig_330  22         2   488   \n",
       "...                                         ...  ..       ...   ...   \n",
       "dDocent_Contig_441023-46  dDocent_Contig_441023  46         2   418   \n",
       "dDocent_Contig_441052-5   dDocent_Contig_441052   5         2   388   \n",
       "dDocent_Contig_441078-18  dDocent_Contig_441078  18         2   398   \n",
       "dDocent_Contig_441136-5   dDocent_Contig_441136   5         2   368   \n",
       "dDocent_Contig_441160-23  dDocent_Contig_441160  23         2   428   \n",
       "\n",
       "                             A1_FREQ      A2_FREQ  \n",
       "dDocent_Contig_123-64     T:0.782787   G:0.217213  \n",
       "dDocent_Contig_237-19     A:0.930328  C:0.0696721  \n",
       "dDocent_Contig_278-64     T:0.973361  C:0.0266393  \n",
       "dDocent_Contig_326-42     C:0.965164  T:0.0348361  \n",
       "dDocent_Contig_330-22     C:0.942623   A:0.057377  \n",
       "...                              ...          ...  \n",
       "dDocent_Contig_441023-46  G:0.973684  T:0.0263158  \n",
       "dDocent_Contig_441052-5   A:0.969072  T:0.0309278  \n",
       "dDocent_Contig_441078-18  C:0.957286  T:0.0427136  \n",
       "dDocent_Contig_441136-5   A:0.932065  C:0.0679348  \n",
       "dDocent_Contig_441160-23  A:0.953271   G:0.046729  \n",
       "\n",
       "[28114 rows x 6 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_123-64</th>\n",
       "      <td>dDocent_Contig_123</td>\n",
       "      <td>64</td>\n",
       "      <td>138/106/0</td>\n",
       "      <td>149.51/82.98/11.51</td>\n",
       "      <td>18.787750</td>\n",
       "      <td>6.680879e-07</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.970073e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_237-19</th>\n",
       "      <td>dDocent_Contig_237</td>\n",
       "      <td>19</td>\n",
       "      <td>225/4/15</td>\n",
       "      <td>211.18/31.63/1.18</td>\n",
       "      <td>186.190600</td>\n",
       "      <td>2.141909e-20</td>\n",
       "      <td>2.141909e-20</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_278-64</th>\n",
       "      <td>dDocent_Contig_278</td>\n",
       "      <td>64</td>\n",
       "      <td>235/5/4</td>\n",
       "      <td>231.17/12.65/0.17</td>\n",
       "      <td>89.268270</td>\n",
       "      <td>2.421640e-06</td>\n",
       "      <td>2.421640e-06</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_326-42</th>\n",
       "      <td>dDocent_Contig_326</td>\n",
       "      <td>42</td>\n",
       "      <td>227/17/0</td>\n",
       "      <td>227.30/16.41/0.30</td>\n",
       "      <td>0.317867</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>7.493871e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_330-22</th>\n",
       "      <td>dDocent_Contig_330</td>\n",
       "      <td>22</td>\n",
       "      <td>227/6/11</td>\n",
       "      <td>216.80/26.39/0.80</td>\n",
       "      <td>145.672900</td>\n",
       "      <td>1.730150e-14</td>\n",
       "      <td>1.730150e-14</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441023-46</th>\n",
       "      <td>dDocent_Contig_441023</td>\n",
       "      <td>46</td>\n",
       "      <td>198/11/0</td>\n",
       "      <td>198.14/10.71/0.14</td>\n",
       "      <td>0.152666</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.737299e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441052-5</th>\n",
       "      <td>dDocent_Contig_441052</td>\n",
       "      <td>5</td>\n",
       "      <td>182/12/0</td>\n",
       "      <td>182.19/11.63/0.19</td>\n",
       "      <td>0.197601</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.391836e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441078-18</th>\n",
       "      <td>dDocent_Contig_441078</td>\n",
       "      <td>18</td>\n",
       "      <td>187/7/5</td>\n",
       "      <td>182.36/16.27/0.36</td>\n",
       "      <td>64.623930</td>\n",
       "      <td>1.872976e-06</td>\n",
       "      <td>1.872976e-06</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441136-5</th>\n",
       "      <td>dDocent_Contig_441136</td>\n",
       "      <td>5</td>\n",
       "      <td>169/5/10</td>\n",
       "      <td>159.85/23.30/0.85</td>\n",
       "      <td>113.507600</td>\n",
       "      <td>9.809512e-13</td>\n",
       "      <td>9.809512e-13</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_441160-23</th>\n",
       "      <td>dDocent_Contig_441160</td>\n",
       "      <td>23</td>\n",
       "      <td>199/10/5</td>\n",
       "      <td>194.47/19.07/0.47</td>\n",
       "      <td>48.383460</td>\n",
       "      <td>1.176736e-05</td>\n",
       "      <td>1.176736e-05</td>\n",
       "      <td>9.999998e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28114 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CHROM  POS OBS(HOM1/HET/HOM2)  \\\n",
       "dDocent_Contig_123-64        dDocent_Contig_123   64          138/106/0   \n",
       "dDocent_Contig_237-19        dDocent_Contig_237   19           225/4/15   \n",
       "dDocent_Contig_278-64        dDocent_Contig_278   64            235/5/4   \n",
       "dDocent_Contig_326-42        dDocent_Contig_326   42           227/17/0   \n",
       "dDocent_Contig_330-22        dDocent_Contig_330   22           227/6/11   \n",
       "...                                         ...  ...                ...   \n",
       "dDocent_Contig_441023-46  dDocent_Contig_441023   46           198/11/0   \n",
       "dDocent_Contig_441052-5   dDocent_Contig_441052    5           182/12/0   \n",
       "dDocent_Contig_441078-18  dDocent_Contig_441078   18            187/7/5   \n",
       "dDocent_Contig_441136-5   dDocent_Contig_441136    5           169/5/10   \n",
       "dDocent_Contig_441160-23  dDocent_Contig_441160   23           199/10/5   \n",
       "\n",
       "                            E(HOM1/HET/HOM2)   ChiSq_HWE         P_HWE  \\\n",
       "dDocent_Contig_123-64     149.51/82.98/11.51   18.787750  6.680879e-07   \n",
       "dDocent_Contig_237-19      211.18/31.63/1.18  186.190600  2.141909e-20   \n",
       "dDocent_Contig_278-64      231.17/12.65/0.17   89.268270  2.421640e-06   \n",
       "dDocent_Contig_326-42      227.30/16.41/0.30    0.317867  1.000000e+00   \n",
       "dDocent_Contig_330-22      216.80/26.39/0.80  145.672900  1.730150e-14   \n",
       "...                                      ...         ...           ...   \n",
       "dDocent_Contig_441023-46   198.14/10.71/0.14    0.152666  1.000000e+00   \n",
       "dDocent_Contig_441052-5    182.19/11.63/0.19    0.197601  1.000000e+00   \n",
       "dDocent_Contig_441078-18   182.36/16.27/0.36   64.623930  1.872976e-06   \n",
       "dDocent_Contig_441136-5    159.85/23.30/0.85  113.507600  9.809512e-13   \n",
       "dDocent_Contig_441160-23   194.47/19.07/0.47   48.383460  1.176736e-05   \n",
       "\n",
       "                          P_HET_DEFICIT  P_HET_EXCESS  \n",
       "dDocent_Contig_123-64      1.000000e+00  3.970073e-07  \n",
       "dDocent_Contig_237-19      2.141909e-20  1.000000e+00  \n",
       "dDocent_Contig_278-64      2.421640e-06  1.000000e+00  \n",
       "dDocent_Contig_326-42      1.000000e+00  7.493871e-01  \n",
       "dDocent_Contig_330-22      1.730150e-14  1.000000e+00  \n",
       "...                                 ...           ...  \n",
       "dDocent_Contig_441023-46   1.000000e+00  8.737299e-01  \n",
       "dDocent_Contig_441052-5    1.000000e+00  8.391836e-01  \n",
       "dDocent_Contig_441078-18   1.872976e-06  1.000000e+00  \n",
       "dDocent_Contig_441136-5    9.809512e-13  1.000000e+00  \n",
       "dDocent_Contig_441160-23   1.176736e-05  9.999998e-01  \n",
       "\n",
       "[28114 rows x 8 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_out = os.path.join(analysis_dir,'loci_out.csv')\n",
    "frq_out = os.path.join(analysis_dir,'frq_out.csv')\n",
    "hardy_out = os.path.join(analysis_dir,'hardy_out.csv')\n",
    "\n",
    "loci_df.to_csv(path_or_buf=loci_out)\n",
    "frq_df.to_csv(path_or_buf=frq_out)\n",
    "hardy.to_csv(path_or_buf=hardy_out)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    28114.000000\n",
       "mean        13.328822\n",
       "std         43.110074\n",
       "min          2.791440\n",
       "25%          5.885430\n",
       "50%          7.888365\n",
       "75%         11.459975\n",
       "max       1784.860000\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    2.811400e+04\n",
       "mean     3.309517e+03\n",
       "std      2.453001e+04\n",
       "min      0.000000e+00\n",
       "25%      4.515400e+02\n",
       "50%      9.979610e+02\n",
       "75%      2.563517e+03\n",
       "max      2.331460e+06\n",
       "Name: QUAL, dtype: float64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.QUAL.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyMAAAIJCAYAAAC2k+IhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3de5xkZX3v+89XEBBQbscLitmDRIStqAFJFE+4yNGgEkUdj5hEUbxEt6AiujWIERSzPQFBASM5GplEkj0oHvWgeDtyd1QUNGhE7iMbxQsCg1wV+J0/1mqtFN09PTPV/UxVf96vV70WtdbzW89TPWXZ3171rCdVhSRJkiQttAe0HoAkSZKkxckwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKa2LD1ADR/klwLPARY2XgokiRJmlxLgFuravs1LTSMTLaHPOhBD9p655133rr1QCRJkjSZLrvsMu688861qjWMTLaVO++889YXX3xx63FIkiRpQu22225ccsklK9em1jkjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWpiw9YD0GRa8o4vtB7Cglv5/ue2HoIkSdJY8cqIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJamKswkiSbZK8OslnklyV5M4kq5JcmORVSR4w1H5JkprlsXyWvg5KclGS2/o+zk2y/yztN0hyWJJL+3HdlOSsJHvMUvOgJEcnuTzJXUl+keSTSXZeu5+QJEmSND42bD2ANfRi4CPADcA5wHXAw4EXAh8Dnp3kxVVVQ3X/Dnx2mvP9YLpOkhwHHA5cD3wU2Ag4EDgzyaFVdfJQ+wDLgaXA5cDJwNbAS4Dzk7yoqj43VLMx8FXg6cB3gA8Bj+5f43OTPKOqvrXan4gkSZI0psYtjFwBPA/4QlXdN7UzyRHARcCL6ILJp4fqvldVR82lg/5KxuHA1cDuVXVzv/9Y4GLguCSfr6qVA2UH0gWRFcC+VXVXX3MKcCHw0SRnV9WvB2reQhdEzgBeMvV6kpxOF5w+nmSXwdcpSZIkTZKx+ppWVZ1dVWcO/4JeVT8DTumf7r2O3byu375vKoj0fawEPgxsDLxyqOb1/fbIqSDS13wbOB14KF1YAX53JWWqn/8++Hr6KygXAP8V2GsdX4skSZK03hqrMLIav+2390xz7JFJ/jrJEf32ibOc5xn99kvTHPviUBuSbALsAdxBFyJWWwPsAPwBcEVVXTvHGkmSJGmijNvXtKaVZEPg5f3T6ULEM/vHYM25wEFVdd3Avs2ARwG3VdUN05znyn6748C+HYANgGuqarogNF3N4/rtFdO0n6lmRkkunuHQTnOplyRJklqYlCsj7weeAJxVVV8e2H8H8F5gN2Cr/rEX3eT3vYGv9QFkyhb9dtUM/Uzt37JBjSRJkjRRxv7KSJI30k04/xHwssFjVfUL4G+HSs5P8iy6ieV/Arya7k5WY6uqdptuf3/FZNcFHo4kSZI0J2N9ZSTJIXRB4ofAPlV101zq+q9Tfax/uufAoakrElswvan9tzSokSRJkibK2IaRJG8GTqJbK2Sf/o5aa+KX/fZ3X9OqqtuBnwCbJ9l2mprH9tvBuR5XA/cCj+nnrsyl5vJ+O9OckOlqJEmSpIkylmEkyduBE4Dv0QWRX6zFaZ7ab68Z2n92v91vmppnD7Whv5XvCmBT4E/nUkMXYK4Ddkyy/RxrJEmSpIkydmEkybvoJqxfTLfA4I2ztN01yf1eY5J9gcP6p6cNHZ5ar+SdSbYaqFkCvAG4Gzh1qOYj/faY/la/UzW7063C/ksGFmLsV4if6ufvB8eY5Pl0oeaHwHkzvTZJkiRp3I3VBPYkBwHvofta1AXAG7v1A/+TlVW1rP/v44HHJlkBXN/veyK/X7/jXVW1YrC4qlYkOZ5uhfRLk5wBbEQXKrYGDh1afR1gOd3K70uB7yY5E9imr9kAeE1V3TpUczywf1/zrSRfo1t75MV0dwE72NXXJUmSNMnGKowAU19p2gB48wxtzgOW9f/9CeAFwO50X316IPBz4JPAyVU13SKFVNXhSb5PdyXktcB9wCXAsVX1+WnaV5KX0n1d62DgUOAu4HzgmOHA09fcneSZwDuAl9JdqbkV+Czw7qr64cw/BkmSJGn8jVUYqaqjgKPWoP0/Af+0ln0t4/ehZi7t76Gbx3LCGtTcQXfr4eHbD0uSJEkTb+zmjEiSJEmaDIYRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNTFWYSTJNkleneQzSa5KcmeSVUkuTPKqJNO+niR7JDkryU19zaVJ3pxkg1n62j/Juf35b0vyrSQHrWZ8ByW5qG+/qq/ff5b2GyQ5rB/Pnf34zkqyx9x/KpIkSdJ4GqswArwY+CjwJ8C3gA8CnwaeAHwM+GSSDBYkeT5wPrAn8BngZGAj4ARg+XSdJDkEOLM/72l9n48EliU5boaa44BlwLZ9+9OAXYAz+/MNt0/f//H9eE7ux7cncH4/bkmSJGlibdh6AGvoCuB5wBeq6r6pnUmOAC4CXgS8kC6gkOQhdMHgXmDvqvpOv/9dwNnA0iQHVtXygXMtAY4DbgKeUlUr+/3vAb4NHJ7k01X1jYGaPYDDgauB3avq5n7/scDFwHFJPj91rt6BwFJgBbBvVd3V15wCXAh8NMnZVfXrdfyZSZIkSeulsboyUlVnV9WZg0Gk3/8z4JT+6d4Dh5YCDwWWTwWRvv1dwJH909cPdXMwsDFw8mB46APG3/VPXzdUM/X8fVNBpK9ZCXy4P98rh2qm+j1yKoj0Nd8GTu/HvRRJkiRpQo1VGFmN3/bbewb2PaPffmma9ucDdwB7JNl4jjVfHGqzVjVJNgH26Pu/YA36kSRJkibGuH1Na1pJNgRe3j8dDASP67dXDNdU1T1JrgUeDzwGuGwONTckuR3YLsmmVXVHks2ARwG3VdUN0wzvyn6748C+HYANgGuq6p77l0xbM6MkF89waKe51EuSJEktTMqVkffTTTY/q6q+PLB/i367aoa6qf1brkXNFkPb+ehjyxmOS5IkSWNv7K+MJHkj3eTxHwEvazycJqpqt+n291dMdl3g4UiSJElzMtZXRvpb5n4I+CGwT1XdNNRk+CrGsKn9t6xFzaqh7Xz0ccsMxyVJkqSxN7ZhJMmbgZOAH9AFkZ9N0+zyfnu/uRf9PJPt6Sa8XzPHmm2BzYDrq+oOgKq6HfgJsHl/fNhj++3gHJSr6W43/Jh+HHOpkSRJkibKWIaRJG+nW7Twe3RB5BczND273+43zbE9gU2BFVV19xxrnj3UZq1q+lv5ruj7/9M16EeSJEmaGGMXRvoFC99Pt5jgvlV14yzNzwBuBA5M8pSBc2wCHNM//chQzanA3cAh/QKIUzVbAUf0T08Zqpl6/s6+3VTNEuAN/flOHaqZ6veYfjxTNbsDLwF+Sb94oyRJkjSJxmoCe5KDgPfQfcXpAuCNSYabrayqZQBVdWuS19CFknOTLKdbWf15dLfwPYNugcHfqaprk7wNOBH4TpLTgd/QLUC4HfCBwdXX+5oVSY4H3gJcmuQMYCO6ULE1cOjQ6usAy+lWi18KfDfJmcA2fc0GwGuq6tY1/iFJkiRJY2KswgjdHA/ofll/8wxtzgOWTT2pqs8m2Qt4J/AiYBPgKrrgcGJV1fAJquqkJCuBt9KtX/IAuknyR1bVP0/XaVUdnuT7dFdCXgvcB1wCHFtVn5+mfSV5Kd3XtQ4GDgXuoluM8ZiqWjHzj0GSJEkaf2MVRqrqKOCotaj7OvCcNaw5EzhzDWuWMRCE5tD+Hrq5LyesST+SJEnSJBi7OSOSJEmSJoNhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNTHSMJLkgaM8nyRJkqTJNeorIz9J8n8l+cMRn1eSJEnShBl1GHkA8Dbg8iRfTfKiJBuMuA9JkiRJE2DUYeSRwF8BFwD7Ap8Erk/yviRLRtyXJEmSpDE20jBSVb+pqn+rqr2BnYAPAhsCfwNcleSsJM9P4sR5SZIkaZGbt1BQVVdU1eHAo/j91ZL9gP8HuC7JUUkeOV/9S5IkSVq/zfsViqr6DfAF4DPAT4HQfZ3rb4Frk3wwycbzPQ5JkiRJ65d5DSNJnprkVLoQcgKwGXAi8GTgYOBy4FC6r3NJkiRJWkQ2HPUJkzwYeBnw18AT6K6EfBf4B+DfqurOvumlST4BfAlYCrx+1GORJEmStP4aaRhJ8k/A/wlsCtwNfAL4h6q6aLr2VXVvknOBZ4xyHJIkSZLWf6O+MvJK4GrgFODUqrppDjXnAu8Z8TgkSZIkredGHUb2q6qvrElBVX0d+PqIxyFJkiRpPTfqdUbWKIhIkiRJWrxGGkaS7Jvk4zOtH5Lkkf3xvUfZryRJkqTxM+qvaR0K7FRVP53uYFX9NMnTgC3o5opIkiRJWqRGvc7IrsCK1bS5EHjKiPuVJEmSNGZGHUYeRrfA4Wx+3reTJEmStIiNOoysAh69mjaPBm4fcb+SJEmSxsyow8hFwAFJHjHdwX5i+wF9O0mSJEmL2KjDyEnAg4ELkjwvycYASTZO8nzgfGBz4MQR9ytJkiRpzIz0blpV9ZUk7wXeBXwGqCQ3A1sB6R/vraovjbJfSZIkSeNn1FdGqKp3A/sBZwE30d3G9ybgC8Cf9cclSZIkLXKjXmcE+N1K7K7GLkmSJGlGI78yIkmSJElzMS9XRgCSbEo3V2SD6Y5X1XXz1bckSZKk9d/Iw0iSlwFvB3aepVnNR9+SJEmSxsdIA0GSVwAfB+4FLgD+F3DPKPuQJEmSNBlGfXXircDNwP9eVZeN+NySJEmSJsioJ7D/IfApg4gkSZKk1Rl1GLkJuHvE55QkSZI0gUYdRj4P7J0kIz6vJEmSpAkz6jDyN8DGwClJNh/xuSVJkiRNkFFPYP8UcAfwauAvklwJ3DJNu6qqfUfctyRJkqQxMuowsvfAf28GPHmGdjXifiVJkiSNmZGGkaoa9de+JEmSJE0ow4MkSZKkJgwjkiRJkpoYeRhJ8oAkhyb5ZpJVSe4ZOPZHSf4hyY6j7leSJEnSeBlpGEmyEfBV4IPADsCvgcE1R64FDgb+cpT9SpIkSRo/o74y8jZgH+Bo4OHAxwYPVtUtwPnAn61tB0mWJjkpyQVJbk1SSU6boe2S/vhMj+Wz9HNQkouS3NZf4Tk3yf6ztN8gyWFJLk1yZ5KbkpyVZI9Zah6U5Ogklye5K8kvknwyyc5r9lORJEmSxs+ob+37l8DXq+o9AEmmu4XvtcCfr0MfRwJPAm4Drgd2mkPNvwOfnWb/D6ZrnOQ44PD+/B8FNgIOBM5McmhVnTzUPsByYClwOXAysDXwEuD8JC+qqs8N1WxMdxXp6cB3gA8BjwZeDDw3yTOq6ltzeG2SJEnSWBp1GNke+MJq2txE94v62jqMLiRcBewFnDOHmu9V1VFzOXl/JeNw4Gpg96q6ud9/LHAxcFySz1fVyoGyA+mCyApg36q6q685BbgQ+GiSs6vq1wM1b6ELImcAL6mq+/qa0+mC08eT7DK1X5IkSZo0o/6a1l3Alqtp8wdMvyr7nFTVOVV1ZVXN18KJr+u375sKIn2/K4EPAxsDrxyqeX2/PXIqiPQ13wZOBx5KF1aA311Jmernvw8Gjv4KygXAf6ULW5IkSdJEGnUY+R7wrH4i+/0k2YJuvshFI+53dR6Z5K+THNFvnzhL22f02y9Nc+yLQ21IsgmwB3AHXYhYbQ3d5P4/AK6oqmvnWDOjJBdP92BuX2GTJEmSmhj117T+b+BfgX9N8qrBA0m2BE4FtgJOGXG/q/PM/jE4nnOBg6rquoF9mwGPAm6rqhumOc+V/Xbw1sQ7ABsA11TVPfcvmbbmcf32ihnGO12NJEmSNFFGGkaq6n8meSbwCuB5wNR8i+8Aj6f7itOHq+qsUfY7izuA99LNwbim3/dE4Ci6u359LcmTq+r2/tgW/XbVDOeb2j/4VbSFqplRVe023f7+6siuczmHJEmStNBGvuhhVR1Mt5bID+nmSoTuF+KrgFdV1aGj7nOWsfyiqv62qi6pqlv6x/nAs4BvAX8IvHqhxiNJkiTp90YeRgCqallV/RGwObAd8OCq2qWqTp2P/tZU/3WqqTVQ9hw4NHVFYgumN7V/cAL+QtVIkiRJE2VewsiUqrqzqn468DWo9ckv++1mUzv6cf4E2DzJttPUPLbfDs71uBq4F3hMkum+9jZdzeX9dqY5IdPVSJIkSRNlXsPIeu6p/faaof1n99v9pql59lAb+lv5rgA2Bf50LjV0AeY6YMck28+xRpIkSZooIw0jSa6Z4+PqUfY7y3h2TXK/15hkX7rFEwFOGzo8daevdybZaqBmCfAG4G66u4IN+ki/Paa/1e9Uze50q7D/Evj01P5+jZSpfv5+cIxJnk8Xan4InLfaFylJkiSNqVHf2vcBwHSLEW7J7+dB/BT47dp2kOQA4ID+6SP67dOSLOv/+8aqemv/38cDj02ygm7VdujupjW1fse7qmrF4PmrakWS4+lWSL80yRnARnShYmvg0KHV1wGWAy+kW9jwu0nOBLbpazYAXlNVtw7VHA/s39d8K8nX6NYeeTHdXcAOdvV1SZIkTbJR39p3yUzHkvwhcCLdHI0/W4dungwcNLTvMf0D4MfAVBj5BPACYHe6rz49EPg58Eng5KqabpFCqurwJN+nuxLyWuA+4BLg2Kr6/DTtK8lL6b6udTBwKN1q9OcDxwwHnr7m7v42yO8AXkp3peZWutsQv7uqfrj6H4UkSZI0vtJ9Y2iBOuu+wvQD4FNV9TcL1vEileTiXXfdddeLL754wfte8o4vLHifra18/3NbD0GSJGnB7bbbblxyySWXzLT23WwWdAJ7P9n7q3RXAiRJkiQtYi3upnUPv5/rIUmSJGmRWtAwkuR/o5vD8b8Wsl9JkiRJ65+RTmBP8rez9PNo4Pl0d9VyvogkSZK0yI361r5Hreb4rXR3l/r7EfcrSZIkacyMOozsM8P++4CbgR9V1T0j7lOSJEnSGBr1OiOuGC5JkiRpTlrcTUuSJEmSRj6B/Q/WtraqrhvlWCRJkiSt30Y9Z2QlsDZLuhejH4skSZKk9dioA8C/AEuAPYFVwPeAn9Etcvhkutv6nkcXWiRJkiQtYqMOI/8D+AZwAnB0Vd06dSDJQ4CjgZcDf11VV4y4b0mSJEljZNQT2N8PfL+qDh8MIgBVdWtVHQb8R99OkiRJ0iI26jCyJ3DhatpcCOw14n4lSZIkjZlRh5GN6eaHzGbbvp0kSZKkRWzUYeS7wIFJ/mi6g0l2A14CXDLifiVJkiSNmVFPYD8a+BLwzST/CpwP/Bx4ON1Xs/6CLgAdPeJ+JUmSJI2ZkYaRqvr/khwI/CPwCuCggcMBbgZeW1VfG2W/kiRJksbPyBcarKozknwReD6wK93aIqvovpr1uaq6fdR9SpIkSRo/87LqeR84/q1/SJIkSdL9jHoC+3+SZKskj57PPiRJkiSNp5GHkSSbJ/lAkp8BNwLXDhz7kyRnJdl11P1KkiRJGi8jDSNJtgC+ARwG/BS4jG7i+pTvA38KvHSU/UqSJEkaP6O+MvJO4PHAK6pqV+BTgwer6g7gPGDfEfcrSZIkacyMOoy8EPhyVf3LLG1+DDxqxP1KkiRJGjOjDiPbAZeups1tdLf7lSRJkrSIjTqM/Bp42GrabE83sV2SJEnSIjbqMPJtYP8kD57uYJJtgecAF464X0mSJEljZtRh5EPANsBZSXYePNA//xSwCXDiiPuVJEmSNGZGugJ7VX05ydHAu4EfAL8FSHIjsBXdbX7fXlUrRtmvJEmSpPEz8kUPq+poulv3/r/AzcC9QAFnAf9HVR076j4lSZIkjZ+RXhlJsidwa1WdA5wzynNLkiRJmiyjvjJyDvDaEZ9TkiRJ0gQadRi5EbhzxOeUJEmSNIFGHUbOBfYY8TklSZIkTaBRh5EjgccleW+SB4743JIkSZImyEgnsAN/Q3dL3yOAVyX5d+BndHfTGlRV9aoR9y1JkiRpjIw6jLxi4L8f0T+mU4BhRJIkSVrERh1Gth/x+SRJkiRNqHUOI0leDnyvqi6tqh+PYEySJEmSFoFRTGBfBhwwuCPJQUnOHsG5JUmSJE2oUd9Na8oSYK95OrckSZKkCTBfYUSSJEmSZmUYkSRJktSEYUSSJElSE6MKI8OLGkqSJEnSrEa1zshRSY4a3pnk3hnaV1WNeo0TSZIkSWNkVIEg89xekiRJ0oRZ5zBSVc47kSRJkrTGDBKSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpoYuzCSZGmSk5JckOTWJJXktNXU7JHkrCQ3JbkzyaVJ3pxkg1lq9k9ybpJVSW5L8q0kB62mn4OSXNS3X9XX7z9L+w2SHNaP585+fGcl2WP1PwlJkiRpvI1dGAGOBA4Bngz8ZHWNkzwfOB/YE/gMcDKwEXACsHyGmkOAM4EnAKcBHwUeCSxLctwMNccBy4Bt+/anAbsAZ/bnG26fvv/j+/Gc3I9vT+D8ftySJEnSxBrHMHIYsCPwEOD1szVM8hC6YHAvsHdVvaqq3kYXZL4BLE1y4FDNEuA44CbgKVX1hqo6DHgicDVweJKnDdXsARzeH39iVR1WVW8AduvPc1x/3kEHAkuBFcCTq+ptVfUqYJ9+vB9N8uC5/lAkSZKkcTN2YaSqzqmqK6uq5tB8KfBQYHlVfWfgHHfRXWGB+weag4GNgZOrauVAzc3A3/VPXzdUM/X8fX27qZqVwIf7871yqGaq3yP78UzVfBs4vR/30tW+QkmSJGlMjV0YWUPP6LdfmubY+cAdwB5JNp5jzReH2qxVTZJNgD36/i9Yg34kSZKkibFh6wHMs8f12yuGD1TVPUmuBR4PPAa4bA41NyS5HdguyaZVdUeSzYBHAbdV1Q3TjOHKfrvjwL4dgA2Aa6rqnjnWzCjJxTMc2mku9ZIkSVILk35lZIt+u2qG41P7t1yLmi2GtvPRx5YzHJckSZLG3qRfGVkUqmq36fb3V0x2XeDhSJIkSXMy6VdGhq9iDJvaf8ta1Kwa2s5HH7fMcFySJEkae5MeRi7vt/ebe5FkQ2B74B7gmjnWbAtsBlxfVXcAVNXtdOudbN4fH/bYfjs4B+Vqutv3PqYfx1xqJEmSpIky6WHk7H673zTH9gQ2BVZU1d1zrHn2UJu1qulv5bui7/9P16AfSZIkaWJMehg5A7gRODDJU6Z29rfWPaZ/+pGhmlOBu4FDBhcqTLIVcET/9JShmqnn7+zbTdUsAd7Qn+/UoZqpfo/pxzNVszvwEuCXwKdX8/okSZKksTV2E9iTHAAc0D99RL99WpJl/X/fWFVvBaiqW5O8hi6UnJtkOd2K6M+ju4XvGXQLDP5OVV2b5G3AicB3kpwO/IZuAcLtgA9U1TeGalYkOR54C3BpkjOAjehCxdbAoYMLKPaWAy/sz/vdJGcC2/Q1GwCvqapb1+JHJEmSJI2FsQsjwJOBg4b2PaZ/APwYeOvUgar6bJK9gHcCLwI2Aa6iCw4nTreSe1WdlGRlf56X011B+iHdaun/PN2gqurwJN+nuxLyWuA+4BLg2Kr6/DTtK8lL6b6udTBwKHAX3WKMx1TVitX/KCRJkqTxNXZhpKqOAo5aw5qvA89Zw5ozgTPXsGYZsGwN2t8DnNA/JEmSpEVl0ueMSJIkSVpPGUYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkbqT3D8AABIFSURBVCRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUxKIII0lWJqkZHj+boWaPJGcluSnJnUkuTfLmJBvM0s/+Sc5NsirJbUm+leSg1YztoCQX9e1X9fX7r+trliRJktZ3G7YewAJaBXxwmv23De9I8nzg08BdwOnATcCfAycATwdePE3NIcBJwK+A04DfAEuBZUl2qaq3TlNzHHA4cD3wUWAj4EDgzCSHVtXJa/4yJUmSpPGwmMLILVV11OoaJXkIXTC4F9i7qr7T738XcDawNMmBVbV8oGYJcBxdaHlKVa3s978H+DZweJJPV9U3Bmr2oAsiVwO7V9XN/f5jgYuB45J8fupckiRJ0qRZFF/TWkNLgYcCy6eCCEBV3QUc2T99/VDNwcDGwMmD4aEPGH/XP33dUM3U8/dNBZG+ZiXw4f58r1yXFyJJkiStzxZTGNk4yV8lOSLJm5LsM8P8j2f02y9Nc+x84A5gjyQbz7Hmi0Nt1qVGkiRJmhiL6WtajwA+MbTv2iSvrKrzBvY9rt9eMXyCqronybXA44HHAJfNoeaGJLcD2yXZtKruSLIZ8Cjgtqq6YZqxXtlvd5zLC0ty8QyHdppLvSRJktTCYrkyciqwL10g2QzYBfhHYAnwxSRPGmi7Rb9dNcO5pvZvuRY1Wwxt16QPSZIkaaIsiisjVXX00K4fAK9LchvdJPKjgBcs9LhGpap2m25/f8Vk1wUejiRJkjQni+XKyExO6bd7DuwbvooxbGr/LWtRs2pouyZ9SJIkSRNlsYeRX/bbzQb2Xd5v7zdfI8mGwPbAPcA1c6zZtj//9VV1B0BV3Q78BNi8Pz7ssf32fnNQJEmSpEmx2MPIU/vtYLA4u9/uN037PYFNgRVVdfcca5491GZdaiRJkqSJMfFhJMnO/d2rhvcvAaZWOD9t4NAZwI3AgUmeMtB+E+CY/ulHhk53KnA3cEh/3qmarYAj+qenDNVMPX9n325wXG/oz3fqLC9NkiRJGmuLYQL7S+hWQD8f+DHwa2AH4LnAJsBZdKunA1BVtyZ5DV0oOTfJcrqV1Z9HdwvfM4DTBzuoqmuTvA04EfhOktOB39AtoLgd8IHB1df7mhVJjgfeAlya5Axgo368WwOHuvq6JEmSJtliCCPn0IWIPwKeTjd/4xbgQrp1Rz5RVTVYUFWfTbIX8E7gRXSh5Sq64HDicPu+5qQkK4G3Ai+nu+r0Q+DIqvrn6QZWVYcn+T7dlZDXAvcBlwDHVtXn1/F1S5IkSeu1iQ8j/YKG56224f3rvg48Zw1rzgTOXMOaZcCyNamRJEmSJsHEzxmRJEmStH4yjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw0hjSbZL8vEkP01yd5KVST6YZKvWY5MkSZLm04atB7CYJdkBWAE8DPgc8CPgj4E3AfsleXpV/arhECVJkqR545WRtv6BLoi8saoOqKp3VNUzgBOAxwHvazo6SZIkaR4ZRhrpr4o8C1gJfHjo8LuB24GXJdlsgYcmSZIkLQjDSDv79NuvVNV9gweq6tfA14FNgacu9MAkSZKkheCckXYe12+vmOH4lXRXTnYEvjbbiZJcPMOhJ1122WXstttuazfCdXDDT1YteJ+tbbzsTa2HsOCe8KgtWg9BkiQ1dtlllwEsWZtaw0g7U7/FzfRb+9T+Ldehj3vvvPPOVZdccsnKdTjH2tip3/5ogfvVArvk56tt4ntB4PtAv+d7QeD7YBItAW5dm0LDyASoqoW/9DGLqSs169u4tPB8Lwh8H+j3fC8IfB/oP3POSDtTVz5m+p7L1P5bFmAskiRJ0oIzjLRzeb/dcYbjj+23M80pkSRJksaaYaSdc/rts5L8p3+HJA8Gng7cAXxzoQcmSZIkLQTDSCNVdTXwFboJP28YOnw0sBnwiaq6fYGHJkmSJC0IJ7C39d+AFcCJSfYFLgP+hG4NkiuAdzYcmyRJkjSvUlWtx7CoJXk08B5gP2Ab4AbgM8DRVXVzy7FJkiRJ88kwIkmSJKkJ54xIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiFYryXZJPp7kp0nuTrIyyQeTbLWG59m6r1vZn+en/Xm3m6+xa7RG8V5Icm6SmuWxyXy+Bq2bJEuTnJTkgiS39v9mp63luUby2aI2RvVe6P/dZ/o8+Nl8jF2jk2SbJK9O8pkkVyW5M8mqJBcmeVWSNfpd08+FxccV2DWrJDvQrRL/MOBzwI+APwbeBOyX5OlV9as5nGeb/jw7AmcDy4GdgFcCz03ytKq6Zn5ehUZhVO+FAUfPsP+edRqo5tuRwJOA24Dr6f53vMbm4f2khTeS90JvFfDBafbftg7n1MJ4MfARukWbzwGuAx4OvBD4GPDsJC+uOSxs5+fCIlVVPnzM+AC+DBRw6ND+4/v9p8zxPP/Yt//A0P439vu/1Pq1+liw98K53UdP+9fkY63eB/sAjwUC7N3/25+2FucZyfvJx0S8F1YCK1u/Hh9r/T54BvDnwAOG9j+CLpgU8KI5nsvPhUX4cAV2zaj/C8VVdP9HsUNV3Tdw7MF0fwUJ8LCqun2W82wO/AK4D9i2qn49cOwBwDXAf+n78OrIemhU74W+/bnAXlWVeRuwFkSSven+EvqvVfVXa1A3sveT1g9r+17oa1cCVNWSkQ9MTSU5AngfcHJVHbqatn4uLFLOGdFs9um3Xxn8UADoA8XXgU2Bp67mPE8FHgR8fTCI9Oe5j+4vIYP9af0zqvfC7yR5SZJ3JHlLkmcn2Xh0w9V6buTvJ429jZP8VZIjkrwpyT5JNmg9KK2z3/bbuXz91s+FRcowotk8rt9eMcPxK/vtjgt0HrUzH/+Gy4H/AXwAOAu4LsnStRuexoyfCRr2COATdH9F/yDd3MIrk+zVdFRaa0k2BF7eP/3SHEr8XFikDCOazRb9dtUMx6f2b7lA51E7o/w3/Bzd94u3o7tithNdKNkSOD3JfuswTo0HPxM06FRgX7pAshmwC908wyXAF5M8qd3QtA7eDzwBOKuqvry6xvi5sGh5Ny1JC6qqThjadTlwRJKfAifRBZO5/BVN0gSoquE76/0AeF2S24DDgaOAFyz0uLT2kryR7t/uR8DLGg9H6zmvjGg2U3+F2GKG41P7b1mg86idhfg3/Bjd94qf3E9W1OTyM0FzcUq/3bPpKLRGkhwCfAj4IbBPVd00x1I/FxYpw4hmc3m/nen7mY/ttzN9v3PU51E78/5vWFV3AVM3ONhsbc+jseBngubil/3Wz4MxkeTNdFe4f0AXRNZk0Uo/FxYpw4hmc06/fdbwCqr9X66fDtwBfHM15/kmcCfw9OG/ePfnfdZQf1r/jOq9MKMkjwO2ogskN67teTQW5v39pIkwddckb/k+BpK8HTgB+B5dEPnFGp7Cz4VFyjCiGVXV1cBX6CYRvmHo8NF0f636xOD9vpPslOQ/rcJbVbfR3SVlM7rv/g46pD//l11jZP01qvdCku2TbD18/iQPpZvECrC8qlyFfQIkeWD/PthhcP/avJ803mZ6LyTZOcn9rnwkWQKc3D89bf5HqHWR5F10E9YvBvatqhn/oOTngoa56KFm1X9YrAAeRncXpMuAP6G7H/gVwB5V9auB9gUwvKBdkm368+xId8vGi4CdgefTLYi4R/9BpPXUKN4LSV5B9z3wC+n+2nkT8AfAc+i+D/wd4JlV5XeC11NJDgAO6J8+Avgzun/LC/p9N1bVW/u2S4BrgR8PL2i3pu8nrX9G8V5IchTdROfzgR/TXRndAXgusAndbb9fUFW/mdcXo7WW5CBgGXAv3Ve0prsb1sqqWta3X4KfCxpgGNFqJXk08B5gP2AbulVQPwMcXVU3D7WdNoz0x7YG3k33f17bAr8Cvgj8bVVdP5+vQaOxru+FJLvQ/eKxG/BI4CF0v3z8B/BJ4B/9pWP91v/y+O5ZmvzuF4zZfunoj8/5/aT1zyjeC/06Iq8D/ojf39r3Frqv+nyC7i/h/qKyHpvD+wDgvKrau2+/BD8XNMAwIkmSJKkJ54xIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSNIaSLE1yUpILktyapJKcNg/97Jrk35Jcn+TuJD9Pcl6Sl6/ruTccxQAlSZIkLbgjgScBtwHXAzuNuoMkhwAfAm4GvgD8BNgaeALwHOBf1uX8hhFJkiRpPB1GF0KuAvYCzhnlyZM8CzgR+CqwtKp+PXT8gevah1/TkiRJksZQVZ1TVVdWVc21JslLk5yT5JYkdyW5LMmRSTaepvmxwJ3AXwwHkb7/367D8AGvjEiSJEmLQpKPA6+ku5ryaeAW4KnAe4F9kzyzqu7p2z4BeCLwWeCmJPsAuwEFfA84p6ruW9cxGUYkSZKkCZfkFXRB5DPAX1bVnQPHjgLeDbyBbn4IwO799hfAucCeQ6f8fpIXVtVV6zIuv6YlSZIkTb43AfcABw8Gkd57gV8Bfzmw72H99lXAEuC5wBbAjsBpwC7AF5JstC6D8sqIJEmSNMGSbEp3160bgTcnma7Z3cDOA8+nLlpsABxYVd/on9/a39J3J+ApwIuA/7m2YzOMSJIkSZNtKyDAQ+m+jjUXt/Tbnw0EEQCqqpJ8ji6M/DHrEEb8mpYkSZI02Vb12+9WVWZ7DNRc3m9vYXo399sHrcvADCOSJEnSBKuq24D/AB6fZOs5ln0TuB1YkmSzaY4/od9euy5jM4xIkiRJk+94YCPg40m2HD6YZKsku049r6o7gH8CNgGOycBEkyS7AK+gmxB/xroMKmuwRookSZKk9USSA4AD+qePAP4MuAa4oN93Y1W9daD9h4H/BtwEfBm4Dtga2J7u1r2nVtXrBto/BDgPeDLwLeDrwMOBF9J9PevNVTV1K+C1ew2GEUmSJGn8DKwPMpMfV9WSoZr9gdfRTTzfki6YXAd8BTitqn401H5z4G+AFwP/hW5F9ouA46rqK+v8GgwjkiRJklpwzogkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQm/n+B095YZUMTngAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 260,
       "width": 401
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25909, 28114)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.SUM_DEPTH <5000]),len(loci_df[loci_df.SUM_DEPTH >=500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1224, 28103)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.MEAN_DEPTH > 30]),len(loci_df[loci_df.MEAN_DEPTH >= 3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "118"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 10]) - len(loci_df[loci_df.QUAL >= 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2747, 1534, 884)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL <  200]), len(loci_df[loci_df.QUAL < 100]), len(loci_df[loci_df.QUAL < 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(52, 364)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " len(loci_df[loci_df.Fis <= -0.5]), len(loci_df[loci_df.MAF < 0.02])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) &  \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.MEAN_DEPTH >= 3) & \n",
    "                  (df.MEAN_DEPTH < 30) & \n",
    "                  (df.QUAL >= 200) & \n",
    "                  (df.MAF >= 0.02) &  \n",
    "                  (df.Fis > -0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23884, 25)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/good_snps\n",
      "\t--positions /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 23884 out of a possible 28114 Sites\n",
      "Run Time = 20.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_bad_remove_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"stage1_positions.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "snps_gz = snps + \".gz\"\n",
    "!$bgzip -c {snps} > {snps_gz}\n",
    "!$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make 012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/good_snps.recode.vcf\n",
      "\t--012\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 23884 out of a possible 23884 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--012"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get coverage per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/good_snps.recode.vcf\n",
      "\t--depth\n",
      "\t--out /data/gpfs/assoc/denovo/PHHA/ddocent/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 244 out of 244 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 23884 out of a possible 23884 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PH_AS_10</td>\n",
       "      <td>23089</td>\n",
       "      <td>11.10270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PH_AS_11</td>\n",
       "      <td>21646</td>\n",
       "      <td>7.03973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PH_AS_12</td>\n",
       "      <td>22343</td>\n",
       "      <td>11.92390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PH_AS_1</td>\n",
       "      <td>22326</td>\n",
       "      <td>12.13930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PH_AS_2</td>\n",
       "      <td>23398</td>\n",
       "      <td>10.37370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       INDV  N_SITES  MEAN_DEPTH\n",
       "0  PH_AS_10    23089    11.10270\n",
       "1  PH_AS_11    21646     7.03973\n",
       "2  PH_AS_12    22343    11.92390\n",
       "3   PH_AS_1    22326    12.13930\n",
       "4   PH_AS_2    23398    10.37370"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = os.path.join(analysis_dir, \"good_snps.recode.vcf.idepth\")\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    244.000000\n",
       "mean       9.396647\n",
       "std        2.079223\n",
       "min        4.394690\n",
       "25%        7.953965\n",
       "50%        9.401110\n",
       "75%       10.909300\n",
       "max       19.157400\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove unnecessary files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm snps*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *miss*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
