{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SNP filtering using vcftools   \n",
    "\n",
    "## Filter rabbit_phylo using the same metrics as filtering without outgroup\n",
    "### Do not remove individuals\n",
    "\n",
    "\n",
    "use preferred conda env  \n",
    "**Packages needed**: vcftools, bgzip, tabix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import ipyparallel as ipp\n",
    "import os\n",
    "from os import environ\n",
    "import gzip\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import glob\n",
    "import re\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools = \"vcftools\"\n",
    "bcftools = \"bcftools\"\n",
    "bgzip = \"bgzip\"\n",
    "tabix = \"tabix\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "root = '/data/gpfs/home/tfaske/d/rabbit/full/phylo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/tfaske/rabbit/full/phylo\n"
     ]
    }
   ],
   "source": [
    "cd $root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp vcf/rabbit_phylo.vcf.gz filtering/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = os.path.join(root,'filtering')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/tfaske/rabbit/full/phylo/filtering\n"
     ]
    }
   ],
   "source": [
    "cd $analysis_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_phylo.vcf.gz'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file = os.path.join(analysis_dir, \"rabbit_phylo.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter first round \n",
    "    missing data, MAF, biallelic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_phylo.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--thin 100\n",
      "\t--max-missing 0.7\n",
      "\t--out rabbit_miss70_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 32343 out of a possible 1428667 Sites\n",
      "Run Time = 173.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--max-missing 0.7 \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $'rabbit_miss70_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"%s.recode.vcf\" % '/data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2'\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter snps further \n",
    "this needs to be done after removing individuals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Quality for Each Site\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 32343 out of a possible 32343 Sites\n",
      "Run Time = 4.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_filtered_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    \n",
    "    hardy_files = !ls {filedir}/{prefix}.hwe\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_files = !ls {filedir}/{prefix}.l* | grep -v log\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    frq_files = !ls {filedir}/{prefix}.frq* | grep -v count\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:45: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/pandas/core/indexing.py:822: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=6, step=1)\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(analysis_dir,'rabbit_miss70_thin100_MAF2.recode.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_out = os.path.join(analysis_dir,'loci_out.csv')\n",
    "frq_out = os.path.join(analysis_dir,'frq_out.csv')\n",
    "hardy_out = os.path.join(analysis_dir,'hardy_out.csv')\n",
    "\n",
    "loci_df.to_csv(path_or_buf=loci_out)\n",
    "frq_df.to_csv(path_or_buf=frq_out)\n",
    "hardy.to_csv(path_or_buf=hardy_out)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     32343.000000\n",
       "mean       6232.457008\n",
       "std       15061.472790\n",
       "min         863.000000\n",
       "25%        1760.000000\n",
       "50%        2528.000000\n",
       "75%        4474.500000\n",
       "max      145573.000000\n",
       "Name: SUM_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    32343.000000\n",
       "mean       946.781829\n",
       "std        142.420981\n",
       "min          6.075910\n",
       "25%        999.000000\n",
       "50%        999.000000\n",
       "75%        999.000000\n",
       "max        999.000000\n",
       "Name: QUAL, dtype: float64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.QUAL.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZEAAAD4CAYAAAAtrdtxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAXo0lEQVR4nO3de7SddX3n8ffHBBC8ESSlaYINakYnthUxIi7bGYsjBOwYnHEsTFtSZUynwiwdXWsM2lVslbWkFy/MKIqaMThqQLyQwThMZFjt+AeXoAgETDlyKYkIEVC8DQj9zh/7F9gczkn2eXL2PueQ92utZ53n+T633/6dJJ88z+/Ze6eqkCSpi6fMdAMkSXOXISJJ6swQkSR1ZohIkjozRCRJnc2f6QaM2qGHHlpLly6d6WZI0pxy7bXX/rCqFo6v73MhsnTpUrZs2TLTzZCkOSXJHRPVvZ0lSerMEJEkdWaISJI6M0QkSZ0ZIpKkzgwRSVJnhogkqTNDRJLU2dBCJMlTk1yd5DtJtib5i1Y/IslVScaSXJhk/1Y/oC2PtfVL+451ZqtvS3J8X31lq40lWTus1yJJmtgw37H+IHBsVf00yX7AN5N8HXgH8KGq2pDk48BpwHnt5/1V9fwkJwPnAL+fZDlwMvAi4NeAbyT5Z+0cHwVeA2wHrkmysapuGtYLWrr2a8M69G7d/oHXzsh5JWlPhnYlUj0/bYv7tamAY4GLW309cFKbX9WWaetfnSStvqGqHqyq24Ax4Og2jVXVrVX1ELChbStJGpGhjokkmZfkOuAeYDPwPeBHVfVw22Q7sLjNLwbuBGjrfww8u78+bp/J6hO1Y02SLUm27Ny5czpemiSJIYdIVT1SVUcCS+hdObxwmOfbTTvOr6oVVbVi4cInfAilJKmjkTydVVU/Aq4AXgEcnGTXWMwSYEeb3wEcDtDWPwu4t78+bp/J6pKkERnm01kLkxzc5g+kNwB+M70weUPbbDVwSZvf2JZp6/9PVVWrn9ye3joCWAZcDVwDLGtPe+1Pb/B947BejyTpiYb5dNYiYH2SefTC6qKqujTJTcCGJO8Hvg18um3/aeCzScaA++iFAlW1NclFwE3Aw8DpVfUIQJIzgMuAecC6qto6xNcjSRpnaCFSVdcDL5mgfiu98ZHx9f8H/LtJjnU2cPYE9U3Apr1urCSpE9+xLknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqbGghkuTwJFckuSnJ1iRva/X3JtmR5Lo2ndi3z5lJxpJsS3J8X31lq40lWdtXPyLJVa1+YZL9h/V6JElPNMwrkYeBd1bVcuAY4PQky9u6D1XVkW3aBNDWnQy8CFgJfCzJvCTzgI8CJwDLgVP6jnNOO9bzgfuB04b4eiRJ4wwtRKrqrqr6Vpv/CXAzsHg3u6wCNlTVg1V1GzAGHN2msaq6taoeAjYAq5IEOBa4uO2/HjhpOK9GkjSRkYyJJFkKvAS4qpXOSHJ9knVJFrTaYuDOvt22t9pk9WcDP6qqh8fVJzr/miRbkmzZuXPnNLwiSRKMIESSPB34EvD2qnoAOA94HnAkcBfwt8NuQ1WdX1UrqmrFwoULh306SdpnzB/mwZPsRy9APldVXwaoqrv71n8SuLQt7gAO79t9SasxSf1e4OAk89vVSP/2kqQRGObTWQE+DdxcVR/sqy/q2+z1wI1tfiNwcpIDkhwBLAOuBq4BlrUnsfanN/i+saoKuAJ4Q9t/NXDJsF6PJOmJhnkl8krgj4AbklzXau+m93TVkUABtwN/AlBVW5NcBNxE78mu06vqEYAkZwCXAfOAdVW1tR3vXcCGJO8Hvk0vtCRJIzK0EKmqbwKZYNWm3exzNnD2BPVNE+1XVbfSe3pLkjQDfMe6JKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKmzoYVIksOTXJHkpiRbk7yt1Q9JsjnJLe3nglZPknOTjCW5PslRfcda3ba/JcnqvvpLk9zQ9jk3SYb1eiRJTzTMK5GHgXdW1XLgGOD0JMuBtcDlVbUMuLwtA5wALGvTGuA86IUOcBbwcuBo4KxdwdO2eUvffiuH+HokSeMMLUSq6q6q+lab/wlwM7AYWAWsb5utB05q86uAC6rnSuDgJIuA44HNVXVfVd0PbAZWtnXPrKorq6qAC/qOJUkagZGMiSRZCrwEuAo4rKruaqt+ABzW5hcDd/bttr3VdlffPkF9ovOvSbIlyZadO3fu1WuRJD1m6CGS5OnAl4C3V9UD/evaFUQNuw1VdX5VraiqFQsXLhz26SRpnzHUEEmyH70A+VxVfbmV7263omg/72n1HcDhfbsvabXd1ZdMUJckjcgwn84K8Gng5qr6YN+qjcCuJ6xWA5f01U9tT2kdA/y43fa6DDguyYI2oH4ccFlb90CSY9q5Tu07liRpBOYP8divBP4IuCHJda32buADwEVJTgPuAN7Y1m0CTgTGgJ8DbwKoqvuSvA+4pm33l1V1X5t/K/AZ4EDg622SJI3IQCGS5Der6oapHLiqvglM9r6NV0+wfQGnT3KsdcC6CepbgN+YSrskSdNn0NtZH0tydZK3JnnWUFskSZozBgqRqvod4A/oDXBfm+TzSV4z1JZJkma9gQfWq+oW4M+AdwH/Ejg3yXeT/JthNU6SNLsNFCJJfivJh+i96/xY4F9X1T9v8x8aYvskSbPYoE9n/VfgU8C7q+oXu4pV9f0kfzaUlkmSZr1BQ+S1wC+q6hGAJE8BnlpVP6+qzw6tdZKkWW3QMZFv0Hsvxi4HtZokaR82aIg8tap+umuhzR80nCZJkuaKQUPkZ+O+JOqlwC92s70kaR8w6JjI24EvJvk+vXeh/yrw+0NrlSRpThgoRKrqmiQvBF7QStuq6pfDa5YkaS6YygcwvgxY2vY5KglVdcFQWiVJmhMG/QDGzwLPA64DHmnlXV9JK0naRw16JbICWN4+aVeSJGDwp7NupDeYLknSowa9EjkUuCnJ1cCDu4pV9bqhtEqSNCcMGiLvHWYjJElz06CP+P5dkl8HllXVN5IcBMwbbtMkSbPdoB8F/xbgYuATrbQY+OqwGiVJmhsGHVg/HXgl8AA8+gVVvzKsRkmS5oZBQ+TBqnpo10KS+fTeJyJJ2ocNGiJ/l+TdwIHtu9W/CPzP4TVLkjQXDBoia4GdwA3AnwCb6H3fuiRpHzbo01n/BHyyTZIkAYN/dtZtTDAGUlXPnfYWSZLmjEFvZ62g9ym+LwN+BzgX+B+72yHJuiT3JLmxr/beJDuSXNemE/vWnZlkLMm2JMf31Ve22liStX31I5Jc1eoXJtl/wNciSZomA4VIVd3bN+2oqg8Dr93Dbp8BVk5Q/1BVHdmmTQBJlgMnAy9q+3wsybwk84CPAicAy4FT2rYA57RjPR+4HzhtkNciSZo+g97OOqpv8Sn0rkx2u29V/X2SpQO2YxWwoaoeBG5LMgYc3daNVdWtrR0bgFVJbgaOBf5922Y9vY9mOW/A80mSpsGgn531t33zDwO3A2/seM4zkpwKbAHeWVX303sH/JV922xvNYA7x9VfDjwb+FFVPTzB9k+QZA2wBuA5z3lOx2ZLksYb9Oms352m850HvI/eIP376IXTm6fp2JOqqvOB8wFWrFjhmyQlaZoMejvrHbtbX1UfHOQ4VXV33zE/CVzaFncAh/dtuqTVmKR+L3BwkvntaqR/e0nSiEzl6aw/pXfLaDHwH4GjgGe0aSBJFvUtvp7el10BbAROTnJAkiOAZcDVwDXAsvYk1v70Bt83tm9YvAJ4Q9t/NXDJoO2QJE2PQcdElgBHVdVPoPeoLvC1qvrDyXZI8gXgVcChSbYDZwGvSnIkvdtZt9N79ztVtTXJRcBN9MZcTq+qR9pxzgAuo/fR8+uqams7xbuADUneD3wb+PSAr0WSNE0GDZHDgIf6lh9qtUlV1SkTlCf9h76qzgbOnqC+id7HrIyv38pjT3BJkmbAoCFyAXB1kq+05ZPoPVYrSdqHDfp01tlJvk7v3eoAb6qqbw+vWZKkuWDQgXWAg4AHquojwPY2AC5J2ocN+vW4Z9EbyD6zlfZjD5+dJUl68hv0SuT1wOuAnwFU1feZwqO9kqQnp0FD5KH23owCSPK04TVJkjRXDBoiFyX5BL13ib8F+AZ+QZUk7fP2+HRWkgAXAi8EHgBeAPx5VW0ectskSbPcHkOkqirJpqr6TcDgkCQ9atDbWd9K8rKhtkSSNOcM+o71lwN/mOR2ek9ohd5Fym8Nq2GSpNlvtyGS5DlV9Y/A8bvbTpK0b9rTlchX6X167x1JvlRV/3YUjZIkzQ17GhNJ3/xzh9kQSdLcs6cQqUnmJUna4+2sFyd5gN4VyYFtHh4bWH/mUFsnSZrVdhsiVTVvVA2RJM09U/koeEmSHscQkSR1ZohIkjozRCRJnRkikqTODBFJUmeGiCSpM0NEktTZ0EIkybok9yS5sa92SJLNSW5pPxe0epKcm2QsyfVJjurbZ3Xb/pYkq/vqL01yQ9vn3PYNjJKkERrmlchngJXjamuBy6tqGXB5WwY4AVjWpjXAedALHeAset9ncjRw1q7gadu8pW+/8eeSJA3Z0EKkqv4euG9ceRWwvs2vB07qq19QPVcCBydZRO97TDZX1X1VdT+9r+dd2dY9s6qurKoCLug7liRpREY9JnJYVd3V5n8AHNbmFwN39m23vdV2V98+QV2SNEIzNrDeriBG8vHySdYk2ZJky86dO0dxSknaJ4w6RO5ut6JoP+9p9R3A4X3bLWm13dWXTFCfUFWdX1UrqmrFwoUL9/pFSJJ6Rh0iG4FdT1itBi7pq5/antI6Bvhxu+11GXBckgVtQP044LK27oEkx7Snsk7tO5YkaUT29KVUnSX5AvAq4NAk2+k9ZfUB4KIkpwF3AG9sm28CTgTGgJ8DbwKoqvuSvA+4pm33l1W1a7D+rfSeADsQ+HqbJEkjNLQQqapTJln16gm2LeD0SY6zDlg3QX0L8Bt700ZJ0t7xHeuSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZ4aIJKkzQ0SS1JkhIknqzBCRJHVmiEiSOjNEJEmdGSKSpM4MEUlSZzMSIkluT3JDkuuSbGm1Q5JsTnJL+7mg1ZPk3CRjSa5PclTfcVa37W9JsnomXosk7ctm8krkd6vqyKpa0ZbXApdX1TLg8rYMcAKwrE1rgPOgFzrAWcDLgaOBs3YFjyRpNGbT7axVwPo2vx44qa9+QfVcCRycZBFwPLC5qu6rqvuBzcDKUTdakvZlMxUiBfzvJNcmWdNqh1XVXW3+B8BhbX4xcGffvttbbbL6EyRZk2RLki07d+6crtcgSfu8+TN03t+uqh1JfgXYnOS7/SurqpLUdJ2sqs4HzgdYsWLFtB1XkvZ1M3IlUlU72s97gK/QG9O4u92mov28p22+Azi8b/clrTZZXZI0IiMPkSRPS/KMXfPAccCNwEZg1xNWq4FL2vxG4NT2lNYxwI/bba/LgOOSLGgD6se1miRpRGbidtZhwFeS7Dr/56vqfyW5BrgoyWnAHcAb2/abgBOBMeDnwJsAquq+JO8Drmnb/WVV3Te6lyFJGnmIVNWtwIsnqN8LvHqCegGnT3KsdcC66W6jJGkws+kRX0nSHGOISJI6M0QkSZ0ZIpKkzgwRSVJnhogkqTNDRJLUmSEiSerMEJEkdWaISJI6M0QkSZ0ZIpKkzgwRSVJnhogkqTNDRJLUmSEiSerMEJEkdWaISJI6M0QkSZ0ZIpKkzgwRSVJnhogkqbP5M90A7dnStV+bsXPf/oHXzti5Jc1+XolIkjozRCRJnc3521lJVgIfAeYBn6qqD8xwk55UZupWmrfRpLlhTl+JJJkHfBQ4AVgOnJJk+cy2SpL2HXP9SuRoYKyqbgVIsgFYBdw0o63SXpvJhwmkJ6NhXd3P9RBZDNzZt7wdePn4jZKsAda0xZ8m2TbF8xwK/LBTC0fLdk4v2zm9bOf0G7itOWevz/XrExXneogMpKrOB87vun+SLVW1YhqbNBS2c3rZzullO6ffbGjrnB4TAXYAh/ctL2k1SdIIzPUQuQZYluSIJPsDJwMbZ7hNkrTPmNO3s6rq4SRnAJfRe8R3XVVtHcKpOt8KGzHbOb1s5/SyndNvxtuaqprpNkiS5qi5fjtLkjSDDBFJUmeGyB4kWZlkW5KxJGtHcL7Dk1yR5KYkW5O8rdUPSbI5yS3t54JWT5JzW/uuT3JU37FWt+1vSbK6r/7SJDe0fc5Nkr1o77wk305yaVs+IslV7dgXtgceSHJAWx5r65f2HePMVt+W5Pi++rT0fZKDk1yc5LtJbk7yitnYn0n+c/ud35jkC0meOlv6M8m6JPckubGvNvQ+nOwcU2znX7ff/fVJvpLk4K591eX3MWg7+9a9M0klOXSm+3MgVeU0yURvsP57wHOB/YHvAMuHfM5FwFFt/hnAP9D7SJe/Ata2+lrgnDZ/IvB1IMAxwFWtfghwa/u5oM0vaOuubtum7XvCXrT3HcDngUvb8kXAyW3+48Cftvm3Ah9v8ycDF7b55a1fDwCOaP09bzr7HlgP/Ic2vz9w8GzrT3pvnL0NOLCvH/94tvQn8C+Ao4Ab+2pD78PJzjHFdh4HzG/z5/S1c8p9NdXfx1Ta2eqH03tQ6A7g0Jnuz4H+bOztAZ7ME/AK4LK+5TOBM0fchkuA1wDbgEWttgjY1uY/AZzSt/22tv4U4BN99U+02iLgu331x203xbYtAS4HjgUubX9gf9j3F/bR/mt/MV7R5ue37TK+T3dtN119DzyL3j/OGVefVf3JY5++cEjrn0uB42dTfwJLefw/zkPvw8nOMZV2jlv3euBzE/XBnvqqy5/vqbYTuBh4MXA7j4XIjPbnniZvZ+3eRB+rsnhUJ2+XxC8BrgIOq6q72qofAIe1+cnauLv69gnqXXwY+C/AP7XlZwM/qqqHJzj2o+1p63/ctp9q+6fqCGAn8N/Tu+32qSRPY5b1Z1XtAP4G+EfgLnr9cy2zrz/7jaIPJztHV2+m9z/zLu3s8ud7YElWATuq6jvjVs3m/jREZqskTwe+BLy9qh7oX1e9/0bM6LPZSX4PuKeqrp3JdgxgPr3bBudV1UuAn9G7jH/ULOnPBfQ+PPQI4NeApwErZ7JNUzGKPtzbcyR5D/Aw8Llpa9Q0SXIQ8G7gz0d1zun6nRkiuzcjH6uSZD96AfK5qvpyK9+dZFFbvwi4Zw9t3F19yQT1qXol8LoktwMb6N3S+ghwcJJdb2LtP/aj7WnrnwXc26H9U7Ud2F5VV7Xli+mFymzrz38F3FZVO6vql8CX6fXxbOvPfqPow8nOMSVJ/hj4PeAP2j+eXdp5L1P/fQzqefT+A/Gd9ndqCfCtJL/aoZ1D78/H2dv7YU/mid7/Ym9tv9xdA2wvGvI5A1wAfHhc/a95/IDYX7X51/L4QberW/0QemMBC9p0G3BIWzd+0O3EvWzzq3hsYP2LPH7g8a1t/nQeP/B4UZt/EY8f3LyV3sDmtPU98H+BF7T597a+nFX9Se/Tp7cCB7XjrAf+02zqT544JjL0PpzsHFNs50p6Xw+xcNx2U+6rqf4+ptLOcetu57ExkRntzz2+jr09wJN9ovdkxD/Qe1rjPSM432/Tu8S8HriuTSfSu796OXAL8I2+Pyyh98Vc3wNuAFb0HevNwFib3tRXXwHc2Pb5b+xhAHCANr+Kx0Lkue0P8Fj7C3dAqz+1LY+19c/t2/89rS3b6Huyabr6HjgS2NL69KvtL9ys60/gL4DvtmN9lt4/brOiP4Ev0Bur+SW9q7vTRtGHk51jiu0cozd2sOvv08e79lWX38eg7Ry3/nYeC5EZ689BJj/2RJLUmWMikqTODBFJUmeGiCSpM0NEktSZISJJ6swQkSR1ZohIkjr7/8TXmr0nU08VAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3057, 27743)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.SUM_DEPTH > 10000]),len(loci_df[loci_df.SUM_DEPTH >=1500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 10]) - len(loci_df[loci_df.QUAL >= 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5663, 1020, 74)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL <  999]), len(loci_df[loci_df.QUAL < 500]), len(loci_df[loci_df.QUAL < 200])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(496, 0)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " len(loci_df[loci_df.Fis <= -0.5]), len(loci_df[loci_df.MAF < 0.02])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) &  \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.SUM_DEPTH >= 1500) & \n",
    "                  (df.SUM_DEPTH < 10000) & \n",
    "                  (df.QUAL >= 500) & \n",
    "                  (df.MAF >= 0.02) &  \n",
    "                  (df.Fis > -0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23832, 25)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/rabbit_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/good_snps\n",
      "\t--positions /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 23832 out of a possible 32343 Sites\n",
      "Run Time = 31.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"stage1_positions.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "snps_gz = snps + \".gz\"\n",
    "!$bgzip -c {snps} > {snps_gz}\n",
    "!$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make 012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/good_snps.recode.vcf\n",
      "\t--012\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 23832 out of a possible 23832 Sites\n",
      "Run Time = 4.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--012"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get coverage per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/good_snps.recode.vcf\n",
      "\t--depth\n",
      "\t--out /data/gpfs/home/tfaske/d/rabbit/full/phylo/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
      "After filtering, kept 588 out of 588 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 23832 out of a possible 23832 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EN_AH_10</td>\n",
       "      <td>23832</td>\n",
       "      <td>7.08010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EN_AH_11</td>\n",
       "      <td>23832</td>\n",
       "      <td>6.58195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EN_AH_12</td>\n",
       "      <td>23832</td>\n",
       "      <td>7.93291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EN_AH_13</td>\n",
       "      <td>23832</td>\n",
       "      <td>7.25231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EN_AH_14</td>\n",
       "      <td>23832</td>\n",
       "      <td>4.78760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       INDV  N_SITES  MEAN_DEPTH\n",
       "0  EN_AH_10    23832     7.08010\n",
       "1  EN_AH_11    23832     6.58195\n",
       "2  EN_AH_12    23832     7.93291\n",
       "3  EN_AH_13    23832     7.25231\n",
       "4  EN_AH_14    23832     4.78760"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = os.path.join(analysis_dir, \"good_snps.recode.vcf.idepth\")\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    588.000000\n",
       "mean       5.594407\n",
       "std        1.527881\n",
       "min        1.316550\n",
       "25%        4.609550\n",
       "50%        5.792695\n",
       "75%        6.734057\n",
       "max        9.327290\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove unnecessary files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm rabbit_miss*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
